---
title: "fisher_test_chrom_4"
author: "JoeColgan"
date: "10 July 2018"
output: html_document
---

```{r}
## Construct table:
## Columns: Coding and noncoding
## Rows: 1% and 99% (rest of the genome)
total_snps <- 1292860

## With the total number of SNPs:
## 1) introns:
total_intron_snps <- 740511

## 2) intergenic regions:
total_intergenic_snps <- 228691

## 3) coding regions:
total_coding_snps <- 35467

## 4) noncoding regions:
total_noncoding_snps <- total_snps - total_coding_snps
```

Calculate the variant types within top 1% of SNPs with strong signatures of EHH:

```{r}
## Within regions under selection:
total_selected_snps <- 10555

## 2) intergenic regions:
intergenic_selected_snps <- 1860

## 3a) intronic regions:
intronic_selected_snps <- 5358

## 3b) intronic regions:
intronic_selected_snps_noncoding <- 163
intronic_selected_snps_coding <- intronic_selected_snps - intronic_selected_snps_noncoding 

## 4) coding regions: 
coding_selected_snps <- 370

```

Calculate the variant types within rest of genome (99% of SNPs) without strong signatures of EHH:

```{r}
## Within regions NOT under selection:
total_non_selected_snps <- total_snps - total_selected_snps

## 2) intergenic regions:
intergenic_non_selected_snps <- total_intergenic_snps - intergenic_selected_snps

## 3) intronic regions:
intronic_non_selected_snps <- total_intron_snps - intronic_selected_snps

## Correct for SNPs found within introns of noncoding genes:
intronic_non_selected_noncoding_snps <- 26964
intronic_non_selected_coding_snps <- intronic_non_selected_snps - intronic_non_selected_noncoding_snps

## 4) coding regions: 
coding_non_selected_snps <- total_coding_snps - coding_selected_snps

```

Calculate a Chi-squared test:

```{r}
## Construct:
chisq.test(rbind(c(coding_selected_snps, coding_non_selected_snps), 
                 c(intronic_selected_snps_coding, intronic_non_selected_coding_snps)))
        
```

Comparison of SNPs within genic regions in comparison to nongenic regions:

```{r}
## Total genomewide SNPs within genic regions:
total_genic_snps <- 819092

## Selected SNPs within genic regions:
genic_selected_snps <- 6212

## SNPs within genic regions not under selectio:
genic_non_selected_snps <- total_genic_snps - genic_selected_snps

## Total SNPs within intergenic regions:
total_intergenic_snps <- 473768

## Selected SNPs within intergenic regions:
intergenic_selected_snps <- 4343

## SNPs within intergenic regions not under selectio:
intergenic_non_selected_snps <- total_intergenic_snps - intergenic_selected_snps

```

Perform a Chi-squared test:

```{r}
## Construct:
chisq.test(rbind(c(genic_selected_snps, genic_non_selected_snps), 
                 c(intergenic_selected_snps, intergenic_non_selected_snps)))
        
```


Fisher'sExact Test for Count Data


data: rbind(c(381, 384), c(11905, 71730))
p-value < 2.2e-16
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
5.166575 6.916962
sample estimates:
odds ratio 
5.977947
