---
title: "plot_manhattan_raw_nsl_go_terms"
author: "JoeColgan"
date: "23 June 2018"
output: html_document
---

```{r}
## Load libraries:
libraries <- c("devtools", "ggplot2", "PopGenome", "qqman", 
               "reshape", "ggpubr", "zoo", "stringr", "dplyr")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}
```

Read in data

```{r}
## Plot raw nsl values:
## Use dataframe list (zoo.files):
## Reformat chromosome name:
#raw_nsl_values <- read.table("./input/combined_genome_wide_new_norm_files.txt", header=FALSE)
raw_nsl_values <- read.table("./input/nsl_output/combined_nsl_input_for_calculating_top_one.txt",
                             header = FALSE)
```

Update the name of each chromosome for plotting:

```{r}
## Extract abbreviated chromosome name:
raw_nsl_values$V1 <- gsub("NC_015762.1", 1, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015763.1", 2, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015764.1", 3, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015765.1", 4, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015766.1", 5, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015767.1", 6, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015768.1", 7, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015769.1", 8, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015770.1", 9, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015771.1", 10, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015772.1", 11, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015773.1", 12, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015774.1", 13, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015775.1", 14, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015776.1", 15, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015777.1", 16, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015778.1", 17, raw_nsl_values$V1)
raw_nsl_values$V1 <- gsub("NC_015779.1", 18, raw_nsl_values$V1)

## Subset columns for plotting:
raw_nsl_values_df <- raw_nsl_values[, c(1,2,4)]
raw_nsl_values_df$SNP<- rownames(raw_nsl_values_df)
raw_nsl_values_df$V4 <- abs(raw_nsl_values_df$V4)

## Update column names:
colnames(raw_nsl_values_df) <- c("CHR", "BP", "P", "SNP")
```


```{r}
## Make sure that 'CHR' is a numeric value:
raw_nsl_values_df$CHR <- as.numeric(raw_nsl_values_df$CHR)

input_for_manhattan_rw <- raw_nsl_values_df %>% 
  
  # Compute chromosome size
  group_by(CHR) %>% 
  summarise(chr_len = max(BP)) %>% 
  
  # Calculate cumulative position of each chromosome
  mutate(tot = cumsum(chr_len) - chr_len) %>%
  select(-chr_len) %>%
  
  # Add this info to the initial dataset
  left_join(raw_nsl_values_df, ., by = c("CHR" = "CHR")) %>%
  
  # Add a cumulative position of each SNP
  arrange(CHR, BP) %>%
  mutate( BPcum = BP + tot)

```


```{r}
## Generate plot:
axisdf <- input_for_manhattan_rw %>% group_by(CHR) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )

## Need to create dataframe containing two columns:
## Column 1: Chromosome (CHR) name  - created as a character value
## Column 2: center, which consists of the chromosome size

## Sort by chromosome number:
#axisdf$CHR<- sort(as.numeric(axisdf$CHR))
#axisdf <- axisdf[order(axisdf$CHR), ]
#axisdf$CHR <- as.character(axisdf$CHR)
```


```{r, message = FALSE}
manhattan_raw_plot<- ggplot(input_for_manhattan_rw,
                            aes(x = BPcum,
                                y = P)) +
            # Show all points
            #geom_point( aes(color=CHR), alpha=0.8, size=1.3) +
            geom_point( aes(color = as.factor(CHR)),
                        alpha = 0.4,
                        size = 1.3) +
            scale_color_manual(values = rep(c("blue", "orange"), 22 )) +
            
            # custom X axis:
            scale_x_continuous( label = axisdf$CHR,
                                breaks = axisdf$center ) +
            scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
            geom_hline(yintercept = c(2.56),
                       linetype = "dashed",
                       colour = "black") +
            xlab("Chromosome") +
            ylab("-log10(P)") +
            # Custom the theme:
            theme_bw() +
            theme( 
              legend.position = "none",
              panel.border = element_blank(),
              panel.grid.major.x = element_blank(),
              panel.grid.minor.x = element_blank()
    )

## Increase the size of the font:
manhattan_raw_plot <- manhattan_raw_plot +
                theme(axis.text = element_text(size = 15),
                axis.title = element_text(size = 15,
                                          face ="bold"),
                axis.text.y = element_text(size = 12),
                axis.text.x = element_text(size = 10))

## Print to console:
manhattan_raw_plot

## Save image
ggsave(manhattan_raw_plot,
       file = "results/manhattan_plot.png",
       dpi = 600,
       height = 10,
       width = 15)
```

Plot alongside a barchart of GO terms enriched within nSL-annotated SNPs:

```{r}
## Read table in for combined plot:
bp_go_terms_df <- read.csv(file="./test_new_output/BP_all.tsv",
                                   header=TRUE, sep="\t")
## Add status column:
bp_go_terms_df$status <- "Biological Processes"

mf_go_terms_df <- read.csv(file="./test_new_output/MF_all.tsv",
                                   header=TRUE, sep ="\t")

## Add status column:
mf_go_terms_df$status <- "Molecular Function"

cc_go_terms_df <- read.csv(file="./test_new_output/CC_all.tsv",
                                   header=TRUE, sep = "\t")

## Add status column:
cc_go_terms_df$status <- "Cellular Component"

## Combine terms and filter for significance:
combined_go_terms_df<- rbind(bp_go_terms_df,
                                mf_go_terms_df,
                                cc_go_terms_df)

## Filter by significance:
combined_go_terms_df_filtered <- subset(combined_go_terms_df, 
                                        weight_ks_adjusted < 0.05)

## Log transform adjusted p values:
combined_go_terms_df_filtered$log10 <- -log10(combined_go_terms_df_filtered$weight_ks_adjusted)

## Create a new name term for plotting:
combined_go_terms_df_filtered$new_go_terms <- paste(combined_go_terms_df_filtered$Term, " (", 
                                                    combined_go_terms_df_filtered$Annotated, ")", sep = "")

## Generate barchart:
bar_plot <- ggbarplot(combined_go_terms_df_filtered, x = "new_go_terms", y = "log10",
          #position = position_dodge(),
          fill = "status",# change fill color by mpg_level
          color = "white",            # Set bar border colors to white
          palette = "jco",# jco journal color palett. see ?ggpar
          sort.val = "asc",          # Sort the value in descending order
          sort.by.groups = TRUE,     # Don't sort inside each group
          x.text.angle = 45,          # Rotate vertically x axis texts,
          ylab = "-log10(p)",
          xlab = "Gene ontology term",
          legend.title = "Gene ontology",
          #label = combined_go_terms_df$Annotated,
          #lab.pos = "out",
          lab.col = "black",
          lab.size = 4,
          lab.vjust = 0.5,
          lab.hjust = 1,
          legend = "top",
          rotate = TRUE,
          ggtheme = theme_minimal()
          )

## Make the axes labels bigger:
bar_plot <- bar_plot +
                theme(axis.text=element_text(size=15),
                axis.title=element_text(size=15,face="bold"),
                axis.text.y = element_text(size=10),
                axis.text.x = element_text(size=10))

## Update colours for plotting:
bar_plot<- bar_plot + 
                scale_fill_manual(values = c("orange", 
                                             "light blue", 
                                             "light grey"))

## Print to console:
bar_plot + geom_hline(yintercept = 1.301, linetype="dashed", colour="black")
```

Plot enrichment of interproscan domains:

```{r}
## Read in interpro results for plotting:
ipr_input <- read.table("./input/interproscan_sorted_by_sig.database.txt", header = FALSE)

## Update the column names:
colnames(ipr_input) <- c("ipr_accession", 
                         "term",
                         "annotated",
                         "bonferroni_adjusted_pvalue")

## Subset significant terms:
ipr_input_sig <- subset(ipr_input, bonferroni_adjusted_pvalue < 0.05)

## Check class and make sure is numeric:
class(ipr_input_sig$bonferroni_adjusted_pvalue)

## Negative log transform adjusted pvalues:
ipr_input_sig$log10 <- -log10(ipr_input_sig$bonferroni_adjusted_pvalue)

## Create new name for plot:
ipr_input_sig$new_term <- paste(ipr_input_sig$term, " (", ipr_input_sig$annotated, ")", sep = "")

ipr_input_sig$new_term <- gsub("_", " ", ipr_input_sig$new_term)

## Generate plot:
ipr_bar_plot <- ggbarplot(ipr_input_sig, x = "new_term", y = "log10",
                        fill = "light grey",           # change fill color by mpg_level
                          color = "white",            # Set bar border colors to white
                          palette = "jco",            # jco journal color palett. see ?ggpar
                          sort.val = "asc",          # Sort the value in descending order
                          sort.by.groups = TRUE,     # Don't sort inside each group
                          x.text.angle = 45,          # Rotate vertically x axis texts
                          ylab = "-log10(P)",
                          xlab = "Interpro accession term",
                          legend.title = "Interpro term enrichment",
                                  #label = ipr_input_sig$bonferroni_adjusted_pvalue,
                          #lab.pos = "out",
                          lab.col = "black",
                          lab.size = 10,
                          lab.vjust = 0.5,
                          lab.hjust = 1,
                          rotate = TRUE,
                          ggtheme = theme_minimal()
                          )

## Make the axes labels bigger:
ipr_bar_plot <- ipr_bar_plot +
                theme(axis.text=element_text(size=15),
                axis.title=element_text(size=15,face="bold"),
                axis.text.y = element_text(size=10),
                axis.text.x = element_text(size=10))

## Print to console:
ipr_bar_plot + geom_hline(yintercept = 1.301, linetype="dashed", colour="black")

```

```{r}
## Plot together:
ggarrange(bar_plot, ipr_bar_plot,
                    ncol=1, nrow = 2, align = "v")
```

