---
title: "Bombus population genomics"
output: plot_pca.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:  
This script takes a VCF file as input (VCF generated by Freebayes) and using gdsfmt generates a genotype matrix.  
Using this genotypic matrix, SNPRelate performs LD pruning, PCA and isolation by descent to assess relatedness amongst individuals.
The output is a number of ggplots for:
- A scatterplot of first two principal components.  
- A scatterplot for IBD co-efficients for each individual in the dataset.  
These plots are output as .PNG files.  

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("gdsfmt",
               "SNPRelate",
               "ggplot2",
               "ggpubr",
               "lintr",
               "reshape2")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE )
    }
}
## Create output directory:
dir.create("results")
```

Load data and perform PCA:

```{r}
## Plot PCA of larger number of samples (n=46):
## Assign the input name:  
input_vcf <- "input/freebayes_hap0_minQ_1_minaltfrac_0.25_minCov1.Bter_n41.NC_only.minQ20.snps_only.no_hets.sites_low_freq.rare_variants_free.maxMeanDP100.recode.ann.vcf"

## Convert VCF to genome data structure (gds) file format
## gds is a container for storing annotation data and SNP genotypes.
## In this format, each byte encodes up to four SNP genotypes thereby reducing file size and access time.
## The GDS format supports data blocking so that only the subset of data that is being processed needs
## to reside in memory. GDS formatted data is also designed for efficient data access to large datasets.

## Use the name of input to designate the name for the GDS file format.
input_vcf_gds <- gsub(".vcf", ".gds", input_vcf)

## Reformat VCF into GDS format
snpgdsVCF2GDS(input_vcf,
              input_vcf_gds,
              method="biallelic.only")

## Extract genotype information:
input_vcf_gds_genofile <- snpgdsOpen(input_vcf_gds)

## Extract genotype information and generate genotypic matrix:  
input_vcf_gds_genofile_matrix <- read.gdsn(index.gdsn(input_vcf_gds_genofile,
                                                      "genotype"))
                                                      
## For performing PCA, it is recommended to perform LD pruning:
snpset <- snpgdsLDpruning(input_vcf_gds_genofile,
                          ld.threshold = 0.1,
                          autosome.only = FALSE)

## Get all selected SNP ids:
snpset.id <- unlist(unname(snpset))

## Perform PCA and store output in variable:  
pca        <- snpgdsPCA(input_vcf_gds_genofile,
                          num.thread = 20,
                          snp.id = snpset.id,
                          autosome.only = FALSE)

## Calculate percentages for principal components:  
pc_percent      <- pca$varprop*100
## Print rounded up percentages:  
print(round(pc_percent, 2))

## Extract sample names from input:  
sample_id     <- read.gdsn(index.gdsn(input_vcf_gds_genofile,
                                      "sample.id"))

## Read in file containing treatment information. 
## This file should contain 'treatment information' for each sample and be in the same order as each sample. 
## i.e. the first treatment information would be assigned to the first sample, etc., etc.
## Read in table:
pop_code_df <- read.table(file = "input/population_information.txt",
                          header = FALSE)

## Rename columns:
colnames(pop_code_df) <- c("sample",
                           "site_number",
                           "landuse")

## 
pop_code <- pop_code_df$landuse

## Generate a dataframe containing a column for:
## 1) Sample name
## 2) Treatment condition
## 3) First principal component of interest
## 4) Second principal component of interest
tab <- data.frame(sample.id = pca$sample.id,
                  pop = factor(pop_code)[match(pca$sample.id, sample_id)],
                  EV1 = pca$eigenvect[,1],    # the first eigenvector
                  EV2 = pca$eigenvect[,2],    # the second eigenvector
                  stringsAsFactors = FALSE)

## Rename the samples for plotting:  
## Renaming the sample.ids
new_names <- list()

## Create a list:
for (name in sample_id){
        new_names[name] <- paste(strsplit(name, "_")[[1]][5],"_",strsplit(name, "_")[[1]][6], sep="")
}

## Unlist as a character string and update sample ids:
tab$sample.id <- as.character(unlist(new_names))

## Add another column for site_id:
tab$site_id <- pop_code_df$site_number

## Check status of whether there are two or one individual per site:
new_site_id <- vector()

for (i in 1:length(tab$site_id)){
        status <- duplicated(tab$site_id)[i]
        if (status==FALSE){
                print("Not duplicated")
                new_site_id <- c(new_site_id, paste(tab$site_id[i], "A", sep=""))
        }
        else {
                print("Duplicated")
                new_site_id <- c(new_site_id, paste(tab$site_id[i], "B", sep=""))
        }
}

## Update site information: 
tab$site_id <- new_site_id

## Alternatively, to plot with just sample ids and not points:
names_plot<- ggplot(tab,
                    aes(x = tab$EV1,
                        y = tab$EV2,
                        color = tab$pop)) +
        #geom_point(shape = 16, size=4, alpha=0.4, show.legend = T) +
        theme_minimal() +
        xlab(paste("Principal component", 1," ","(",round(pc_percent[1], 2), "%", ")", sep="")) +
        ylab(paste("Principal component", 2," ","(",round(pc_percent[2], 2), "%", ")", sep="")) + 
        geom_text(size=6, position=position_jitter(width=0.04, height=0.04), aes(label=tab$site_id, color=tab$pop)) +
        scale_color_manual(values= c("blue", "orange", "red"))

## Adjust the size of the labels:
names_plot <- names_plot +
              theme(axis.text=element_text(size = 20),
                    axis.title=element_text(size = 20,
                                            face = "bold")) +
                    theme(legend.position = "none")

## Plot:
ggsave(file = "pca_plot_PC1_PC2.pruned.0.1.png",
       width = 10,
       height = 10)

## Plot together:
ggarrange(names_n46_plot, 
          names_plot,
          ncol=2, nrow=1,
          labels = c("A", "B"))

```

Explore relatedness among individuals:

```{r, message = FALSE}
# Estimate IBD coefficients
ibd_mom <- snpgdsIBDMoM(input_vcf_gds_genofile,
                    sample.id = sample_id,
                    snp.id = snpset.id,
                    maf = 0.05,
                    missing.rate = 0.05,
                    num.thread = 20,
                    autosome.only = FALSE)

# Make a data.frame
ibd_mom_coeff <- snpgdsIBDSelection(ibd_mom)


## Alternatively, to plot with just sample ids and not points:
ibd_plot<- ggplot(ibd_mom_coeff,
                    aes(x = ibd_mom_coeff$k0,
                        y = ibd_mom_coeff$k1)) +
        geom_point(shape = 16, size=4, alpha=0.4, show.legend = T) +
        theme_minimal() +
        xlab("K0") +
        ylab("K1")
        #geom_text(size=6, position=position_jitter(width=0.04, height=0.04), aes(label=tab$site_id, color=tab$pop)) +
        #scale_color_manual(values= c("blue", "orange", "red"))

ggsave(file = "ibd_mon_plot.png",
       width = 10,
       height = 10)

set.seed(100)
snp.id <- sample(snpset.id, 2000)  # random 2000 SNPs
ibd_mle <- snpgdsIBDMLE(input_vcf_gds_genofile,
                    sample.id = sample_id,
                    snp.id = snp.id,
                    maf = 0.05,
                    missing.rate = 0.05,
                    num.thread = 20,
                    autosome.only = FALSE)

# Make a data.frame
ibd_coeff_mle <- snpgdsIBDSelection(ibd_mle)

ibd_mle_plot<- ggplot(ibd_coeff_mle,
                    aes(x = k0,
                        y = k1)) +
                        geom_point()
                        
        theme_minimal() +
        xlab("K0") +
        ylab("K1")

## Reporting of kinship:
ggsave(file = "ibd_mle_plot.png",
       width = 10,
       height = 10)

## Identity by state:
ibs <- snpgdsIBS(genofile,
                 num.thread = 20,
                 autosomes.only = FALSE)

## Order samples by population information:
pop.idx <- order(pop_code)

input_for_melt <- ibs$ibs
input_for_melt_melted <- melt(input_for_melt)

## Generate plot:
plot <- ggplot(data = input_for_melt_melted,
               aes(x = Var1,
                   y = Var2,
                   fill = value)) + geom_tile()

## Save plot:
ggsave("correlation_plot_IBS.png",
       width = 10,
       height = 10)
```

We want a means to cycle through the first four principal components:

```{r}
## Define function for plotting principal components:
plot_pcomponents <- function(component_one, component_two, component_three, component_four){
                  tab_n46 <- data.frame(sample.id = pca_n46$sample.id,
                                  pop = factor(pop_code_n46)[match(pca_n46$sample.id, sample_id_n46)],
                                  EV1 = pca_n46$eigenvect[,component_one],    # the first eigenvector
                                  EV2 = pca_n46$eigenvect[,component_two],    # the second eigenvector
                                  stringsAsFactors = FALSE)
                  print(tab_n46)
                  
                                names_n46_plot<- ggplot(tab_n46, aes(x=tab_n46$EV1, y=tab_n46$EV2, color = tab_n46$pop)) +
                                   geom_point(shape=16, size=4, alpha=0.4, show.legend = T) +
                                        theme_minimal() +
                                        xlab(paste("Principal component", component_one," ","(",round(pc_percent_n46[component_one], 2), "%", ")", sep="")) +
                                        ylab(paste("Principal component", component_two," ","(",round(pc_percent_n46[component_two], 2), "%", ")", sep="")) + 
                                        #geom_text(size=6, position=position_jitter(width=0.04, height=0.04), aes(label=tab_n46$site_id, color=tab_n46$pop)) +
                                        scale_color_manual(values= c("blue", "orange", "red"))
                                
                                ## Adjust the size of the labels:
                                names_n46_plot <- names_n46_plot +
                                                 theme(axis.text=element_text(size=15),
                                                 axis.title=element_text(size=15,face="bold")) +
                                                 theme(legend.position="none")
                                
                                print(names_n46_plot)
                                
                                ## SECOND PLOT:
                                tab_n46_2 <- data.frame(sample.id = pca_n46$sample.id,
                                  pop = factor(pop_code_n46)[match(pca_n46$sample.id, sample_id_n46)],
                                  EV1 = pca_n46$eigenvect[,component_one],    # the first eigenvector
                                  EV2 = pca_n46$eigenvect[,component_three],    # the second eigenvector
                                  stringsAsFactors = FALSE)
                  print(tab_n46)
                  
                                names_n46_plot_2 <- ggplot(tab_n46_2, 
                                                           aes(x=tab_n46_2$EV1, 
                                                               y=tab_n46_2$EV2, 
                                                               color = tab_n46_2$pop)) +
                                   geom_point(shape=16, size=4, alpha=0.4, show.legend = T) +
                                        theme_minimal() +
                                        xlab(paste("Principal component", component_one," ","(",round(pc_percent_n46[component_one], 2), "%", ")", sep="")) +
                                        ylab(paste("Principal component", component_three," ","(",round(pc_percent_n46[component_three], 2), "%", ")", sep="")) + 
                                        #geom_text(size=6, position=position_jitter(width=0.04, height=0.04), aes(label=tab_n46$site_id, color=tab_n46$pop)) +
                                        scale_color_manual(values= c("blue", "orange", "red"))
                                
                                ## Adjust the size of the labels:
                                names_n46_plot_2 <- names_n46_plot_2 +
                                                 theme(axis.text=element_text(size=15),
                                                 axis.title=element_text(size=15,face="bold")) +
                                                 theme(legend.position="none")
                                
                                ## Print to console:
                                print(names_n46_plot_2)
                                ggarrange(names_n46_plot, names_n46_plot_2)
                                
                                ## THIRD PLOT:
                                tab_n46_3 <- data.frame(sample.id = pca_n46$sample.id,
                                  pop = factor(pop_code_n46)[match(pca_n46$sample.id, sample_id_n46)],
                                  EV1 = pca_n46$eigenvect[,component_one],    # the first eigenvector
                                  EV2 = pca_n46$eigenvect[,component_four],    # the second eigenvector
                                  stringsAsFactors = FALSE)

                                names_n46_plot_3 <- ggplot(tab_n46_3, 
                                                           aes(x=tab_n46_3$EV1, 
                                                               y=tab_n46_3$EV2, 
                                                               color = tab_n46_3$pop)) +
                                   geom_point(shape=16, size=4, alpha=0.4, show.legend = T) +
                                        theme_minimal() +
                                        xlab(paste("Principal component", component_one," ","(",round(pc_percent_n46[component_one], 2), "%", ")", sep="")) +
                                        ylab(paste("Principal component", component_four," ","(",round(pc_percent_n46[component_four], 2), "%", ")", sep="")) + 
                                        #geom_text(size=6, position=position_jitter(width=0.04, height=0.04), aes(label=tab_n46$site_id, color=tab_n46$pop)) +
                                        scale_color_manual(values= c("blue", "orange", "red"))
                                
                                ## Adjust the size of the labels:
                                names_n46_plot_3 <- names_n46_plot_3 +
                                                 theme(axis.text=element_text(size=15),
                                                 axis.title=element_text(size=15,face="bold")) +
                                                 theme(legend.position="none")
                        
                                
                                ## FOURTH PLOT:
                                tab_n46_4 <- data.frame(sample.id = pca_n46$sample.id,
                                  pop = factor(pop_code_n46)[match(pca_n46$sample.id, sample_id_n46)],
                                  EV1 = pca_n46$eigenvect[,component_two],    # the first eigenvector
                                  EV2 = pca_n46$eigenvect[,component_three],    # the second eigenvector
                                  stringsAsFactors = FALSE)
                  
                                names_n46_plot_4 <- ggplot(tab_n46_4, 
                                                           aes(x=tab_n46_4$EV1, 
                                                               y=tab_n46_4$EV2, 
                                                               color = tab_n46_4$pop)) +
                                   geom_point(shape=16, size=4, alpha=0.4, show.legend = T) +
                                        theme_minimal() +
                                        xlab(paste("Principal component", component_two," ","(",round(pc_percent_n46[component_two], 2), "%", ")", sep="")) +
                                        ylab(paste("Principal component", component_three," ","(",round(pc_percent_n46[component_three], 2), "%", ")", sep="")) + 
                                        #geom_text(size=6, position=position_jitter(width=0.04, height=0.04), aes(label=tab_n46$site_id, color=tab_n46$pop)) +
                                        scale_color_manual(values= c("blue", "orange", "red"))
                                
                                ## Adjust the size of the labels:
                                names_n46_plot_4 <- names_n46_plot_4 +
                                                 theme(axis.text=element_text(size=15),
                                                 axis.title=element_text(size=15,face="bold")) +
                                                 theme(legend.position="none")
                                
                                ## Print to console:
                                print(names_n46_plot_4)
                                
                                ## FIFTH PLOT:
                                tab_n46_5 <- data.frame(sample.id = pca_n46$sample.id,
                                  pop = factor(pop_code_n46)[match(pca_n46$sample.id, sample_id_n46)],
                                  EV1 = pca_n46$eigenvect[,component_two],    # the first eigenvector
                                  EV2 = pca_n46$eigenvect[,component_four],    # the second eigenvector
                                  stringsAsFactors = FALSE)
                  
                                names_n46_plot_5 <- ggplot(tab_n46_5, 
                                                           aes(x=tab_n46_5$EV1, 
                                                               y=tab_n46_5$EV2, 
                                                               color = tab_n46_5$pop)) +
                                   geom_point(shape=16, size=4, alpha=0.4, show.legend = T) +
                                        theme_minimal() +
                                        xlab(paste("Principal component", component_two," ","(",round(pc_percent_n46[component_two], 2), "%", ")", sep="")) +
                                        ylab(paste("Principal component", component_four," ","(",round(pc_percent_n46[component_four], 2), "%", ")", sep="")) + 
                                        #geom_text(size=6, position=position_jitter(width=0.04, height=0.04), aes(label=tab_n46$site_id, color=tab_n46$pop)) +
                                        scale_color_manual(values= c("blue", "orange", "red"))
                                
                                ## Adjust the size of the labels:
                                names_n46_plot_5 <- names_n46_plot_5 +
                                                 theme(axis.text=element_text(size=15),
                                                 axis.title=element_text(size=15,face="bold")) +
                                                 theme(legend.position="none")
                                
                                ## SIXTH PLOT:
                                tab_n46_6 <- data.frame(sample.id = pca_n46$sample.id,
                                  pop = factor(pop_code_n46)[match(pca_n46$sample.id, sample_id_n46)],
                                  EV1 = pca_n46$eigenvect[,component_three],    # the first eigenvector
                                  EV2 = pca_n46$eigenvect[,component_four],    # the second eigenvector
                                  stringsAsFactors = FALSE)
                  
                                names_n46_plot_6 <- ggplot(tab_n46_4, 
                                                           aes(x=tab_n46_6$EV1, 
                                                               y=tab_n46_6$EV2, 
                                                               color = tab_n46_6$pop)) +
                                   geom_point(shape=16, size=4, alpha=0.4, show.legend = T) +
                                        theme_minimal() +
                                        xlab(paste("Principal component", component_three," ","(",round(pc_percent_n46[component_three], 2), "%", ")", sep="")) +
                                        ylab(paste("Principal component", component_four," ","(",round(pc_percent_n46[component_four], 2), "%", ")", sep="")) + 
                                        #geom_text(size=6, position=position_jitter(width=0.04, height=0.04), aes(label=tab_n46$site_id, color=tab_n46$pop)) +
                                        scale_color_manual(values= c("blue", "orange", "red"))
                                
                                ## Adjust the size of the labels:
                                names_n46_plot_6 <- names_n46_plot_6 +
                                                 theme(axis.text=element_text(size=15),
                                                 axis.title=element_text(size=15,face="bold")) +
                                                 theme(legend.position="none")
                                
                                ## Combined plot:
                                ggarrange(names_n46_plot, names_n46_plot_2,
                                          names_n46_plot_3, names_n46_plot_4,
                                          names_n46_plot_5, names_n46_plot_6,
                                          nrow=3, ncol = 2,
                                          labels = c("A", "B", "C", "D", "E", "F"))

}

## Call function:
plot_pcomponents(1,2,3,4)

```
