---
title: "plot_phylogenetic_tree"
author: "JoeColgan"
date: "11 July 2018"
output: html_document
---

The purpose of this script is to take a newick-aligned dataset of protein sequences and plot the phylogeny. 

## Load libraries:

```{r, message = FALSE}
## Load libraries:
libraries <- c("tidyverse",
               "ggtree",
               "rphylopic",
               "ggplot2",
               "EBImage")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}
## Create output directory:
dir.create(path = "./plot")
```

Read in input data in newick format.
Generate plot and save to file.

```{r, message = FALSE}
## Read newick alignment file in and store as variable:
tree <- read.tree("./input/RAxML_bipartitions.putative_homologues_refseq_names_edited_clos_added_dist_bacteria_aligned_100.tree")

## A preliminary tree to check internal nodes
ggtree(tree) + geom_text2(aes(subset = !isTip,
                              label = label),
                          hjust = -.3) + geom_tiplab()


## Save:
ggsave(file = "./results/figures/phylo_plot_n100_bs_plus_clos_and_dist_bacteria.png",
       height = 15,
       width = 20)

data(chiroptera, package="ape")
groupInfo <- split(tree$tip.label, gsub("_\\w+", "", tree$tip.label))
chiroptera <- groupOTU(tree, groupInfo)
ggtree(chiroptera, aes(color=group), layout='circular') + geom_tiplab(size=1, aes(angle=angle))

nwk <- system.file("extdata", "sample.nwk", package="treeio")
tree <- read.tree(nwk)
test <- ggtree(tree) + geom_text2(aes(subset=!isTip, label=node), hjust=-.3) + geom_tiplab()

## Save:
ggsave(file = "phylo_plot_n100_bs_plus_clos_and_dist_bacteria.png",
       height = 15,
       width = 20)

p <- ggtree(tree)
viewClade(p+geom_tiplab(), node=99)

tree <- groupOTU(tree, .node = c(99, 133, 169))
ggtree(tree, aes(color=group, linetype=group))

cls <- list(c1 = c(99),
            c2 = c(133),
            c3 = c(169))
tree <- groupClade(tree, cls)

## Plot phylogenetic tree:
phylo_tree <- ggtree(tree, aes(color = group,
                               linetype = group)) +
        #xlim(0, 2.5) +
        geom_tiplab(size = 4,
                    offset = 0.03) +
        geom_label2(aes(subset = !isTip,
                        label = label),
                    size = 3,
                    color = "darkred",
                    alpha = 1) +
        #geom_hilight(node = 99,
        #             fill = "yellow",
        #             alpha = 0.5) +
        theme_tree2("white") +
        xlab("Branch length (my)") +
        scale_color_manual(values = c("blue",
                                      "orange",
                                      "brown",
                                      "brown")) +
        theme(axis.title = element_text(face = "bold",
                                         size = 15))
        #scale_color_continuous(low = "grey",
        #                       high = "grey",
         #                      name = "Branch length (my)")

## Plot:
plot(phylo_tree) + guides(fill = FALSE)

## Save image:
dir.create(path = "./results")
ggsave(file = "./results/plot_wolbachia.png",
       height = 15,
       width = 20)
```
Run lintr:

```{r, message = FALSE}
lintr::lint(file = "./plot_phylogenetic_tree.Rmd")
```

