---
title: "fisher_test_chrom_4"
author: "JoeColgan"
output: html_document
---

```{r}
## Construct table:
data <- read.table(file = "input/combined_all_nsl_annotated_snps_n46_snp_positions.annotated_top_one_percent.txt",
                   header = FALSE)

## Columns: Coding and noncoding
## Rows: 1% and 99% (rest of the genome)
total_snps <- nrow(data)
## With the total number of SNPs:
## 1) introns:
total_intron_snps <- length(grep("intron", data$V4))
## 2) intergenic regions:
true_intergenic_snps <- length(grep("intergenic", data$V4))
true_updwnstream_snps <- length(grep("stream", data$V4))
total_intergenic_snps <- true_intergenic_snps + true_updwnstream_snps
## 3) coding regions:
missense_snps <- length(grep("missense", data$V4))
synonymous_snps <- length(grep("synonymous", data$V4))
stop_variants <- length(grep("stop", data$V4))
start_variants <- length(grep("start", data$V4))
initiator_variants <- length(grep("initiator", data$V4))
splice_region_variants <- length(grep("splice_region_variant", data$V4))
total_genic_coding <- missense_snps +
                synonymous_snps +
                stop_variants +
                start_variants +
                initiator_variants +
                splice_region_variants
## 4) noncoding regions:
utr_snps <- length(grep("UTR_variant", data$V4))
non_coding <- length(grep("non_coding", data$V4))
total_genic_noncoding <- utr_snps +
                         non_coding +
                         total_intron_snps
```

Calculate the variant types within top 1% of SNPs with strong signatures of EHH:

```{r}
## Within regions under selection:
top_one_percent <- subset(data,
                          V3 > quantile(data$V3, 0.99))

## 1) introns:
special_intron_snps <- length(grep("intron", top_one_percent$V4))
## 2) intergenic regions:
special_intergenic_snps <- length(grep("intergenic", top_one_percent$V4))
special_updwnstream_snps <- length(grep("stream", top_one_percent$V4))
total_special_intergenic_snps <- special_intergenic_snps + special_updwnstream_snps
## 3) coding regions:
special_missense_snps <- length(grep("missense", top_one_percent$V4))
special_synonymous_snps <- length(grep("synonymous", top_one_percent$V4))
special_stop_variants <- length(grep("stop", top_one_percent$V4))
special_start_variants <- length(grep("start", top_one_percent$V4))
special_initiator_variants <- length(grep("initiator", top_one_percent$V4))
special_splice_region_variants_1 <- length(grep("splice_region_variant", top_one_percent$V4))
special_splice_region_variants_2 <- length(grep("splice_region_variant&", top_one_percent$V4))
special_splice_region_variants <- special_splice_region_variants_1 - special_splice_region_variants_2
special_genic_coding <- special_missense_snps +
                special_synonymous_snps +
                special_stop_variants +
                special_start_variants +
                special_initiator_variants +
                special_splice_region_variants
## 4) noncoding regions:
special_utr_snps <- length(grep("UTR_variant", top_one_percent$V4))
special_non_coding <- length(grep("non_coding", top_one_percent$V4))
special_genic_noncoding <- special_utr_snps +
                         special_non_coding +
                         special_intron_snps
```

Calculate the variant types within rest of genome (99% of SNPs) without strong signatures of EHH:

```{r}
## Within regions NOT under selection:
nonspecial_genic_noncoding <- total_genic_noncoding - special_genic_noncoding
nonspecial_genic_coding <- total_genic_coding - special_genic_coding
nonspecial_intron <- total_intron_snps - special_intron_snps
nonspecial_intergenic <- total_intergenic_snps - total_special_intergenic_snps
nonspecial_utr <- total_updwnstream_snps - special_updwnstream_snps
```

Calculate a Chi-squared test:

```{r}
## Construct:
chisq.test(rbind(c(special_genic_coding, nonspecial_genic_coding), 
                 c(special_intron_snps, nonspecial_intron)))
```

Perform a Chi-squared test:

```{r}
## Construct:
special_genic_total <- special_genic_coding + 
                        special_genic_noncoding
       
nonspecial_genic_total <- nonspecial_genic_coding + 
                        nonspecial_genic_noncoding
 
special_intergenic_total <- special_intergenic_snps +
                            special_updwnstream_snps

nonspecial_intergenic_total <- nonspecial_intergenic +
                               nonspecial_utr

chisq.test(rbind(c(special_genic_total, nonspecial_genic_total), 
                 c(special_intergenic_total, nonspecial_intergenic_total)))
        
```


Fisher'sExact Test for Count Data


data: rbind(c(381, 384), c(11905, 71730))
p-value < 2.2e-16
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
5.166575 6.916962
sample estimates:
odds ratio 
5.977947
