---
title: "Bombus population genomics"
output: plot_snp_heatmap.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:
This script reads in a VCF file and generates a genotype matrix.
Using data contained in the genotype matrix, a heatmap is generated showing the location of
major and minor alleles.
The output is a .png plot of the heatmap.

## Load libraries:

```{r, message = FALSE}
libraries <- c("lintr",
               "SNPRelate",
               "reshape",
               "ggplot2")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE)
        }
}

## Create output folder:
dir.create(path = "./results/figures",
           recursive = TRUE)
```


```{r, echo = FALSE}
## Define path to input files:
input_path <- "input/vcf/"
input_vcf <- paste(input_path,
                   "input_bed_file.snp_positions.recode.vcf",
                   sep = "")
## Use the name of input to designate the name for the GDS file format.
input_vcf_gds <- gsub("recode.vcf",
                      "recode.gds",
                      input_vcf)
## Reformat VCF into GDS format
snpgdsVCF2GDS(input_vcf,
              input_vcf_gds,
              method = "biallelic.only")
## Extract genotype information:
input_vcf_gds_genofile <- snpgdsOpen(input_vcf_gds)
## Extract genotype information and generate genotypic matrix:
input_vcf_gds_genofile_matrix <- read.gdsn(index.gdsn(input_vcf_gds_genofile,
                                                      "genotype"))
## Perform PCA and store output in variable:
pca      <- snpgdsPCA(input_vcf_gds_genofile,
                      num.thread = 2,
                      autosome.only = FALSE)
## Calculate percentages for principal components:
pc_percent      <- pca$varprop * 100
## Print rounded up percentages:
print(round(pc_percent, 2))
## Extract sample names from input:
sample.id      <- read.gdsn(index.gdsn(input_vcf_gds_genofile,
                                       "sample.id"))
## Print to console:
snpgdsClose(input_vcf_gds_genofile)
```

Plot heatmap to demonstrate representation of haplotypes within population:

```{r}
## Format change '2' to '1':
input_vcf_gds_genofile_matrix[input_vcf_gds_genofile_matrix == 2] <- 1

## Read in table:
pop_code_df <- read.table(file = "input/sample_info/population_information.txt",
                          header = FALSE,
                          col.names = c("sample",
                                        "site_number",
                                        "landuse",
                                        "original_site_number",
                                        "latitude"))

## Check status of whether there are two or one individual per site:
new_names <- vector()

for (i in seq(pop_code_df$site_number)) {
        status <- duplicated(pop_code_df$site_number)[i]
        if (status == FALSE) {
                print("Not duplicated")
                new_names <- c(new_names,
                                 paste(pop_code_df$site_number[i],
                                       "A",
                                       sep = ""))
        }
        else {
                print("Duplicated")
                new_names <- c(new_names,
                                 paste(pop_code_df$site_number[i],
                                       "B",
                                       sep = ""))
        }
}

##
rownames(input_vcf_gds_genofile_matrix) <- new_names

## Read in genomic coordinates:
positions <- read.table(file = "input/bed_file/input_bed_file.snp_positions.txt",
                        header = FALSE)
positions_to_plot <- positions$V2
## Assign to colnames:
colnames(input_vcf_gds_genofile_matrix) <- positions_to_plot

## Convert into dataframe:
input_vcf_gds_genofile_matrix_subset_df <- as.data.frame(input_vcf_gds_genofile_matrix)
dat <- input_vcf_gds_genofile_matrix[, c(50:342)]
row_order <- hclust(dist(dat))$order # clustering
col_order <- hclust(dist(t(input_vcf_gds_genofile_matrix_subset_df)))$order
# re-order matrix accoring to clustering
dat_new <- dat[row_order, ]
dat_new_df <- as.data.frame(dat_new)
dat_new_df$samples <- rownames(dat_new_df)
## Convert heatmap2 into ggplot2 format:
## Reshape:
dat_new_melted <- melt(dat_new_df,
                       id.vars = "samples")

dat_new_melted_ordered <- dat_new_melted[order(dat_new_melted$variable), ]

## Rename:
colnames(dat_new_melted_ordered) <- c("samples",
                                      "variable",
                                      "value")

## Adjust levels:
dat_new_melted_ordered$samples <- factor(dat_new_melted_ordered$samples,
                                 levels = c(unique(as.character(unlist(dat_new_melted_ordered$samples)))))

dat_new_melted_ordered$variable <- factor(as.character(dat_new_melted_ordered$variable),
                                          levels = c(unique(as.character(unlist(dat_new_melted_ordered$variable)))))

## Plot:
heatmap_plot <- ggplot(dat_new_melted_ordered,
                       aes(variable, samples)) +
                  geom_tile(aes(fill = value), colour = "white") +
                  scale_fill_gradient(low = "beige", high = "bisque4") +
                  ylab("Samples") +
                  xlab("Genomic position (bp)") +
        theme(legend.title = element_text(size = 12),
              axis.text.y = element_text(size = 8),
              legend.text = element_text(size = 12),
              plot.title = element_text(size = 12),
              axis.title = element_text(size = 15,
                                        face = "bold"),
              axis.text.x = element_text(size = 8,
                                         angle = 45,
                                         hjust = 1)) +
        theme(legend.position = "none") +
        scale_x_discrete(breaks = levels(dat_new_melted_ordered$variable)[c(T,
                                                                            rep(F, 20))])
```

Save the heatmap to output:

```{r, message = FALSE}
## Save:
dir.create(path = "results/plots/")
save(heatmap_plot,
     file = "results/plots/wolbachia_snp_plot.RData")
```

Run lintr:

```{r, message = FALSE}
## Run lintr
lintr::lint(file = "plot_snp_heatmap_wolbachia.Rmd")
```
