---
title: "Bombus population genomics"
output: plot_wolbachia_sweep.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:
This script plots nsl score for SNPs identified in a gene of interest, an uncharacterised gene (LOC105666162), which codes for proteins with sequence similarity to proteins coded for by Wolbachia-like bacteria. The gene is located on chromosome 11:
NC_015771.1 (1755259..1765481).
The script takes a tab-delimited output file generated by selscan as input.
The script generates and outputs a scatterplot displaying the genomic position of SNPs for the gene of interest, as well as its nsl score. 

1. Load libraries:

```{r, message = FALSE}
## Load libraries:
libraries <- c("devtools",
               "ggplot2",
               "ggmap",
               "SNPRelate",
               "reshape",
               "ggpubr",
               "gplots",
               "stringr",
               "gggenes")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE)
        }
}
```

2. Load input:

```{r, message = FALSE}
## Read in data:
nsl_input <- "input/nsl/input_for_r_annotated.txt"
nsl_data  <- read.table(file = nsl_input,
                        col.names = c("chrom",
                                     "position",
                                     "nsl_score",
                                     "annotation",
                                     "locus"))
```

3. Subset gene of interest, as well as flanking regions:

```{r, message = FALSE}
## Assign the coordinates of the gene to variables plus 100kb
## up and downstream of the gene:
chromosome <- "NC_015771.1"
start      <- 1749721 - 100000
end        <- 1776527 + 100000

## Subset nsl values for plotting:
nsl_input_subset <- subset(nsl_data,
                           chrom == chromosome &
                            position > start &
                            position < end)

## subset sites above 2.72 (i.e. top 10% of genes) to highlight on plot:
nsl_snps_highlight <- subset(nsl_input_subset,
                            nsl_score > 2.87)
```

4. Generate plot:

```{r, message = FALSE}
## Highlight SNP with hightes nsl score:
nsl_input_subset$high_nsl <- nsl_input_subset$nsl_score == 
        max(nsl_input_subset$nsl_score)

## Generate scatterplot:
nsl_plot <- ggplot(nsl_input_subset,
                   aes(x = position,
                       y = nsl_score,
                       fill = high_nsl)) +
        xlab("Genomic position (bp)") +
        labs(y = expression("|"~italic(n)*S[L]*"| score")) +
        geom_rect(aes(xmin = 1725000,
                      xmax = 1800000,
                      ymin = -Inf,
                      ymax = Inf),
                      fill = "beige",
                      alpha = 0.03,
                      color = NA) +
        geom_hline(yintercept = c(2.87),
                   linetype = "dashed",
                   colour = "purple") +
        scale_x_continuous(labels = scales::comma,
                           limits = c(start, end)) +
        geom_point(size = 3,
                   colour = "black",
                   pch = 21,
                   alpha = 0.5) +
        theme_bw() +
        scale_fill_manual(values = c("black",
                                     "darkorange1")) +
        scale_x_continuous(labels = scales::comma,
                           limits = c(start, end),
                           position = "top") +
        theme_bw() +
        theme(axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title = element_text(size = 15,
                                        face = "bold"),
              axis.title.x = element_blank(),
              legend.position = "none")
```

5. Save image:

```{r, message = FALSE}
save.image(file = "results/plots/wolbachia_sweep_scatterplot.RData")
```

6. Run lintr:

```{r, message = FALSE}
## Run lintr:
lintr::lint(file = "plot_nsl_scatterplot.Rmd")
```