---
title: "Bombus population genomics"
output: make_combined_manhattan_plot.html
authors: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:
This script takes input and generates histogram plots as output.  
The first plot generates a histogram plot of variant type by raw |nsl| score.
The second plot generates a histogram plot of max |nsl| score by gene.

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2",
               "ggpubr",
               "lintr")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        avebiocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
```

Load previously generated Manhattan and stacked bar plot:

```{r, message = FALSE}
## Load libraries:
load(file = "./results/manhattan_plot.RData")
load(file = "./results/figures/nsl_variant.Rdata")
```

Plot together:

```{r, message = FALSE}
ggarrange(manhattan_plot,
          scaled_plot,
          labels = c("A",
                     "B"),
          align = "hv",
          nrow = 2,
          ncol = 1)

## Save image:
ggsave(file = "results/figures/combined_manhattan.png",
       height = 10,
       width = 12)
```

```{r, message = FALSE}
## Run lintr:
lintr::lint(file = "./make_combined_manhattan_plot.Rmd")
```
