---
title: "Bombus population genomics"
output: plot_snp_heatmap.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:
This script reads in a VCF file and generates a genotype matrix.
Using data contained in the genotype matrix, a heatmap is generated showing the location of
major and minor alleles.
The output is a .png plot of the heatmap.

## Load libraries:

```{r, message = FALSE}
libraries <- c("lintr",
               "SNPRelate",
               "reshape",
               "ggplot2")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}

## Create output folder:
dir.create(path = "./results/figures",
           recursive = TRUE)
```


```{r, message = FALSE}
## Define input:
input_path <- "./input/vcf_input/regions_of_interest/"
input_vcf <- paste(input_path,
                   "chrom_one_gene_coordinates.snp_positions.recode.vcf",
                   sep = "")
## Use the name of input to designate the name for the GDS file format.
input_vcf_gds <- gsub("recode.vcf",
                      "recode.gds",
                      input_vcf)
## Reformat VCF into GDS format
snpgdsVCF2GDS(input_vcf,
              input_vcf_gds,
              method = "biallelic.only")
## Extract genotype information:
input_vcf_gds_genofile <- snpgdsOpen(input_vcf_gds)
## Extract genotype information and generate genotypic matrix:  
input_vcf_gds_genofile_matrix <- read.gdsn(index.gdsn(input_vcf_gds_genofile,
                                                      "genotype"))
## Perform PCA and store output in variable:  
pca      <- snpgdsPCA(input_vcf_gds_genofile,
                      num.thread = 2,
                      autosome.only = FALSE)
## Calculate percentages for principal components:  
pc_percent      <- pca$varprop * 100
## Print rounded up percentages:  
print(round(pc_percent, 2))
## Extract sample names from input:  
sample.id      <- read.gdsn(index.gdsn(input_vcf_gds_genofile,
                                       "sample.id"))
## Print to console:
snpgdsClose(input_vcf_gds_genofile)
```

4) Plot heatmap to demonstrate representation of haplotypes within population:

```{r}
## Format change '2' to '1'
input_vcf_gds_genofile_matrix[input_vcf_gds_genofile_matrix == 2] <- 1
## Read in sample names:
names <- scan(paste(input_path,
                    "sample_names.txt",
                    sep = ""),
              as.character())

## Update row names:
rownames(input_vcf_gds_genofile_matrix) <- names
## Read in genomic coordinates:
positions <- read.table(paste(input_path,
                              "chrom_one_snp_positions.txt",
                              sep = ""),
                        header = FALSE)
positions_to_plot <- positions$V2
## Assign to colnames:
colnames(input_vcf_gds_genofile_matrix) <- positions_to_plot

## Convert into dataframe:
input_vcf_gds_genofile_matrix_subset_df <- as.data.frame(input_vcf_gds_genofile_matrix)
dat <- input_vcf_gds_genofile_matrix
row_order <- hclust(dist(dat))$order # clustering
col_order <- hclust(dist(t(dat)))$order
dat_new <- dat[row_order, col_order] # re-order matrix accoring to clustering
dat_new_df <- as.data.frame(dat_new)
dat_new_df$samples <- rownames(dat_new_df)
## Convert heatmap2 into ggplot2 format:
## Reshape:
dat_new_melted <- melt(dat_new,
                       id.vars = "samples")

dat_new_melted_ordered <- dat_new_melted[order(dat_new_melted$X2), ]

## Rename:
colnames(dat_new_melted_ordered) <- c("samples",
                                      "variable",
                                      "value")

## Adjust levels:
dat_new_melted_ordered$samples <- factor(dat_new_melted_ordered$samples,
                                 levels = c(unique(as.character(unlist(dat_new_melted_ordered$samples)))))

dat_new_melted_ordered$variable <- factor(as.character(dat_new_melted_ordered$variable),
                                          levels = c(unique(as.character(unlist(dat_new_melted_ordered$variable)))))
                                
## Plot:
heatmap_plot <- ggplot(dat_new_melted_ordered,
                       aes(variable, samples)) +
                  geom_tile(aes(fill = value), colour = "white") +
                  scale_fill_gradient(low = "black", high = "grey") +
                  ylab("Samples") +
                  xlab("Genomic coordinates (bp)") +
theme(legend.title = element_text(size = 15),
        axis.text.y = element_text(size = 6),
        legend.text = element_text(size = 15),
        plot.title = element_text(size = 15),
        axis.title = element_text(size = 15,
                                face = "bold"),
        axis.text.x = element_text(angle = 45,
                                   hjust = 1)) +
        theme(legend.position = "none")

## Output plot:
ggsave(file = "./results/figures/snp_heatmap.png",
       height = 10,
       width = 10)
```

Save the heatmap to output:

```{r, message = FALSE}
## Save:
dir.create(path = "results/plots/")
save(heatmap_plot,
     file = "results/plots/chrom_1_snp_plot.Rdata")
```

Run lintr:
```{r, message = FALSE}
## Run lintr:
lintr::lint(file = "./plot_snp_heatmap_chrom_1.Rmd")
```
