---
title: "Bombus population genomics"
output: generate_combined_plot_apis.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:
This script takes four tab-delimited files as input:
1) A file containing absolute nsl scores assigned to each SNP in the British population.
2) A file containing absolute iHS scores assigned to each SNP in the British population. Both the nsl and iHS scores were calculated by [selscan](https://github.com/szpiech/selscan).
3) A file containing population-scaled recombination rate as calculaed by [ldhelmet](https://sourceforge.net/projects/ldhelmet/).
4) A file containing nucleotide diversity estimates for sliding windows of 10kb as calculated by [popgenome](https://cran.r-project.org/web/packages/PopGenome/index.html).
The input files are used to generate individual scatterplots using [ggplot2](https://ggplot2.tidyverse.org/).
The script also loads arrow-plots and heatmaps generated as output by other scripts.
The scripts generates a combined multiple graph using [ggpubr](https://rpkgs.datanovia.com/ggpubr/index.html), which is output to file.
Code style is checked using [lintr](https://cran.r-project.org/web/packages/lintr/index.html).

1. Load libraries:  

```{r, message = FALSE}
## Load libraries:
libraries <- c("ggplot2",
               "SNPRelate",
               "reshape2",
               "ggpubr",
               "stringr",
               "gggenes")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE)
        }
}
```

2. Load input files:  
For each population (n = 4) of honeybees, load nucleotide diversity estimates for sliding windows of 10kb.  

```{r, message = FALSE}
## Both nucleotide diversity and Tajima D are in the same input file:
input_path <- "input/amel_nuc_div/"
africa_nuc_div_data_10kb_apis <- read.table(file = paste(input_path,
                                                         "apis_all_chrom_10kb_genome_wide_africa.txt",
                                                         sep = ""),
                                                         header = TRUE)
asia_nuc_div_data_10kb_apis <- read.table(file = paste(input_path,
                                                         "apis_all_chrom_10kb_genome_wide_asia.txt",
                                                         sep = ""),
                                                         header = TRUE)
western_nuc_div_data_10kb_apis <- read.table(file = paste(input_path,
                                                         "apis_all_chrom_10kb_genome_wide_western.txt",
                                                         sep = ""),
                                                         header = TRUE)
eastern_nuc_div_data_10kb_apis <- read.table(file = paste(input_path,
                                                         "apis_all_chrom_10kb_genome_wide_eastern.txt",
                                                         sep = ""),
                                                         header = TRUE)

## Combine into a dataframe for plotting:
combined_df <- cbind(africa_nuc_div_data_10kb_apis[, 1:5],
                     asia_nuc_div_data_10kb_apis[, 4:5],
                     eastern_nuc_div_data_10kb_apis[, 4:5],
                     western_nuc_div_data_10kb_apis[, 4:5])
## Add midpoint coordinates:
combined_df$midpoint <- africa_nuc_div_data_10kb_apis$midpoint

## Update column names:
colnames(combined_df) <- c("chrom",
                           "start",
                           "end",
                           "africa_nd",
                           "africa_td",
                           "asia_nd",
                           "asia_td",
                           "eastern_nd",
                           "eastern_td",
                           "western_nd",
                           "western_td",
                           "midpoint")
```

3. Define region of interest:   
These two regions of interest contain 46 genes.  

```{r, message=FALSE}
## Assign the coordinates of the gene to variables plus 100kb
## up and downstream of the gene:
chrom            <- "NC_037638.1"
region_one_start <- 20537332 - 100000
region_one_end   <- 20616537 + 100000

region_two_start <- 25036596 - 100000
region_two_end   <- 25143386 + 100000
```

4. Subset genomic region to plot nucleotide diversity:  

```{r, message = FALSE}
## Subset region:
nuc_div_subset_one <- subset(combined_df,
                         chrom == "NC_037638.1" &
                         start > 20537332 &
                         end < 20616537)

nuc_div_subset_two <- subset(combined_df,
                         chrom == "NC_037638.1" &
                         start > 25036596 &
                         end < 25143386)

## Calculate mean nucleotide diversity:
nucdiv_median <- median(c(combined_df$africa_nd,
                          combined_df$asia_nd,
                          combined_df$eastern_nd,
                          combined_df$western_nd))

## Subset nucleotide diversity measurements:
combined_nd_df      <- combined_df[, c(4, 6, 8, 10, 12)]
combined_nd_df$mean <- rowMeans(combined_nd_df[, c(1:4)])
genome_wide_mean    <- mean(combined_nd_df$mean)
african_nd_mean     <- mean(nuc_div_subset_one$africa_nd)
asian_nd_mean       <- mean(nuc_div_subset_one$asia_nd)
eastern_nd_mean     <- mean(nuc_div_subset_one$eastern_nd)
western_nd_mean     <- mean(nuc_div_subset_one$western_nd)

genome_wide_mean / african_nd_mean
genome_wide_mean / asian_nd_mean
genome_wide_mean / eastern_nd_mean
genome_wide_mean / western_nd_mean

african_nd_mean     <- mean(nuc_div_subset_two$africa_nd)
asian_nd_mean       <- mean(nuc_div_subset_two$asia_nd)
eastern_nd_mean     <- mean(nuc_div_subset_two$eastern_nd)
western_nd_mean     <- mean(nuc_div_subset_two$western_nd)

genome_wide_mean / african_nd_mean
genome_wide_mean / asian_nd_mean
genome_wide_mean / eastern_nd_mean
genome_wide_mean / western_nd_mean

combined_nd_df$mean <- NULL

## Reshape:
combined_nd_melt <- melt(combined_nd_df,
                         id.vars = "midpoint")

## Update column names:
colnames(combined_nd_melt) <- c("midpoint",
                                "population",
                                "nuc_diversity")

## Reformat population name for plotting:
combined_nd_melt$population <- gsub("africa_nd",
                                    "Africa",
                                    combined_nd_melt$population)
combined_nd_melt$population <- gsub("asia_nd",
                                    "Asia",
                                    combined_nd_melt$population)
combined_nd_melt$population <- gsub("eastern_nd",
                                    "Eastern Europe",
                                    combined_nd_melt$population)
combined_nd_melt$population <- gsub("western_nd",
                                    "Western Europe",
                                    combined_nd_melt$population)

## Generate plot:
region_one_plot <- ggplot(combined_nd_melt,
                          aes(x = midpoint,
                              y = nuc_diversity,
                              colour = population)) +
        geom_rect(xmin = 20537332,
                  xmax = 20625795,
                  ymin = -Inf,
                  ymax = Inf,
                  fill = "beige",
                  color = NA,
                  alpha = 0.03) +
        geom_point(colour = "black",
                   size = 0,
                   alpha = 0.75) +
        geom_line() +
        xlab("Genomic coordinates (bp)") +
        ylab("Nucleotide diversity (pi)") +
        geom_hline(yintercept = c(nucdiv_median),
                   linetype = "dashed",
                   colour = "orange") +
        ylim(0, 0.01) +
        scale_x_continuous(limits = c(region_one_start,
                                      region_one_end),
                           #labels = scales::comma,
                           position = "top") +
        scale_color_manual(name = "Population",
                           values = c("blue",
                                      "orange",
                                      "black",
                                      "purple")) +
        theme_bw() +
        theme(axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title = element_text(size = 12,
                                        face = "bold"),
              axis.title.x = element_blank(),
              legend.text = element_text(size = 8,
                                         face = "plain"),
              legend.title = element_text(size = 8,
                                          face = "bold"),
              legend.justification = c(1.1, 1.1),
              legend.position = c(1, 1)
)

## Generate plot for region 2:
region_two_plot <- ggplot(combined_nd_melt,
                          aes(x = midpoint,
                              y = nuc_diversity,
                              colour = population)) +
        geom_rect(xmin = 25036596,
                  xmax = 25143386,
                  ymin = -Inf,
                  ymax = Inf,
                  fill = "beige",
                  color = NA,
                  alpha = 0.03) +
        geom_point(colour = "black",
                   size = 0,
                   alpha = 0.75) +
        geom_line() +
        xlab("Genomic coordinates (bp)") +
        ylab("Nucleotide diversity (pi)") +
        geom_hline(yintercept = c(nucdiv_median),
                   linetype = "dashed",
                   colour = "orange") +
        ylim(0, 0.01) +
        scale_x_continuous(limits = c(region_two_start,
                                      region_two_end),
                           #labels = scales::comma,
                           position = "top") +
        scale_color_manual(name = "Population",
                           values = c("blue",
                                      "orange",
                                      "black",
                                      "purple")) +
        theme_bw() +
        theme(axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title = element_text(size = 12,
                                        face = "bold"),
              axis.title.x = element_blank(),
              axis.title.y = element_blank(),
              axis.text.y = element_blank(),
              axis.ticks.y = element_blank())
```

5. Generate combined plot:  

```{r, message = FALSE}
ggarrange(region_one_plot,
          region_two_plot,
          labels = c("G", "H"),
          ncol = 2,
          nrow = 1,
          align = "hv")
```

6. Load previously generated arrow plot:  

```{r, message = FALSE}
## Load arrow plot:
load(file = "results/plots/chrom_1_arrow_plot_region_1.Rdata")
load(file = "results/plots/chrom_1_arrow_plot_region_2.Rdata")
```

7. Generate combined plots:  

```{r, message=FALSE}
## Update colour scheme for arrow plots:
arrow_plot_one <- arrow_plot_one +
        theme(legend.position = "none") +
        scale_fill_manual(values = c("grey75",
                                     "beige"))
arrow_plot_two <- arrow_plot_two +
        theme(legend.position = "none") +
        scale_fill_manual(values = c("grey75",
                                     "beige"))
## Remove legend from second plot:
region_two_plot <- region_two_plot +
        theme(legend.position = "none")

## Generate a multi-figure plot:
combined_apis_plot <- ggarrange(arrow_plot_one,
                                arrow_plot_two,
                                region_one_plot,
                                region_two_plot,
                                nrow = 2,
                                ncol = 2,
                                labels = c("E",
                                           "G",
                                           "F",
                                           "H"),
                                heights = c(0.2,
                                            1),
                                align = "hv")

## Save object to file:
save(data = combined_apis_plot,
     file = "results/combined_apis_plots.RData")

## Save:
ggsave(file = "results/chrom_one_combined_plot.png",
       width = 8,
       height = 5)

ggsave(file = "results/chrom_one_combined_plot.pdf",
       width = 8,
       height = 5)
```

8. Run lintr:  

```{r, message = FALSE}
## Run lintr:
lintr::lint(file = "generate_combined_plot_apis.Rmd")
```
