---
title: "Bombus population genomics"
output: plot_gggenes_arrows_chrom_2.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction
This script takes a tab-delimited file as input containing the following columns:
1) Chromosome name
2) Start position of chromosome
3) End position of chromosome
4) Gene name
5) Strand
6) Status - if you want to highlight a gene of interest
The script generates an arrow plot and outputs plots as saved R objects (.RData).  

1. Load libraries:  

```{r, message = FALSE}
## Load libraries:
libraries <- c("ggplot2",
               "ggpubr",
               "gggenes",
               "lintr")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}

dir.create(path = "./results/plots",
           recursive = TRUE)
```

2. Load input data:  

```{r, message = FALSE}
## Read in file for plotting:
input <- "./input/arrow_plot_input/region_two_sites_for_gggenes_input.txt"
data <- read.table(file = input,
                   col.names = c("molecule",
                                "start",
                                "end",
                                "gene",
                                "strand"))

data <- subset(data, start > 24936596 &
                       end < 25221884)

## Define start and end for use as limits on x axis:
region_two_start <- 25036596 - 100000
region_two_end <- 25143386 + 100000

## Assign values to start and end positions:
start_position <- data$start[1]
end_position <- data$end[nrow(data)]

data$status <- data$start > 25032154 & data$end <= 25143386

## Turn off scientific notation:
options(scipen = 999)

data$direction <- ifelse(data$strand == "forward", 1, -1)

## Plot:
arrow_plot_two <- ggplot(data, aes(xmin = as.numeric(start),
                              xmax = as.numeric(end),
                              y = strand,
                              fill = status,
                           #   label = gene,
                              forward = direction)) +
        xlim(region_two_start, region_two_end) +
  geom_gene_arrow(arrowhead_height = unit(3, "mm"),
                  arrowhead_width = unit(1, "mm")) +
  #geom_gene_label(align = "left",
#                  min.size = 6) +
  facet_wrap(~ molecule,
             scales = "free",
             nrow = 1,
             shrink = FALSE) +
  scale_fill_brewer(palette = "Blues") +
        theme_genes() +
        theme(axis.title.y = element_blank(),
              axis.text.y = element_blank(),
              axis.title.x = element_blank(),
              axis.text.x = element_blank(),
              axis.ticks.x = element_blank(),
              legend.position = "none")

## Combine plot:
save(arrow_plot_two,
       file = "./results/plots/chrom_1_arrow_plot_region_2.Rdata")
```

3. Run lintr:  

```{r, message = FALSE}
## Run lintr:
lintr::lint(file = "plot_gggene_arrows_chrom_1_region_2.Rmd")
```

