---
title: "Bombus population genomics"
output: run_chisquare_analysis.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:
This script takes in as input a tab-delimited text file containing the following:
1) Chromosome that SNP is located.
2) Genomic position of SNP.
3) |nsl| score of SNP.  
4) Functional annotation of SNP.
The script performs a Chi Squared test to identify enrichment of particular functional variant categories in SNPs with high |nsl| scores (top 1%).  

1. Load libraries:

```{r, message = FALSE}
## Read in data:
data <- read.table(file = "data/annotation_file.txt",
                   header = FALSE)
## Update column names:
colnames(data) <- c("chrom",
                    "position",
                    "nsl_score",
                    "annotation",
                    "locus")
```

2. Group SNPs by functional annotation:

```{r, message = FALSE}
data$annotation_simplified <- data$annotation
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "UTR")] <- "UTR"
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "stream")] <- "Intron"
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "intron")] <- "Intron"
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "intergenic")] <- "Intergenic"
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "non_coding_transcript")] <- "Non-coding RNA"
levels(data$annotation_simplified)[!(levels(data$annotation_simplified) %in% c("UTR",
                                                                               "5kb up/down-stream",
                                                                               "Intron",
                                                                               "Intergenic",
                                                                               "Non-coding RNA"))] <- "Coding"
data$annotation_simplified <- factor(data$annotation_simplified,
                                     levels = rev(c("Non-coding RNA",
                                                    "Coding",
                                                    "UTR",
                                                    "Intergenic",
                                                    "5kb up/down-stream",
                                                    "Intron")))
```

3. Extract SNPs in the top 1% with strongest signatures of selection:

```{r, message = FALSE}
nsl_top_one_percent <- quantile(data$nsl_score,
                                0.99)
## Within regions under selection:
nsl_top_one_percent_df <- subset(data,
                          nsl_score > nsl_top_one_percent)
```

4. Next generate counts for each variant type:

```{r, message = FALSE}
## Get total number of SNPs in dataset:
total_snps <- nrow(data)
## Function to calculate total SNPs for a variant type as well as the number
## of SNPs in the top 1%:
extract_snps <- function(all_snps,
                         snps_of_interest,
                         category_of_snp) {
        total_snps             <- length(grep(category_of_snp,
                                         all_snps$annotation_simplified))
        total_snps_of_interest <- length(grep(category_of_snp,
                                         snps_of_interest$annotation_simplified))
        corrected_snps <- total_snps - total_snps_of_interest
        print(c(corrected_snps, total_snps_of_interest))
}
## Run function:
run_chiseq_test <- function(data_1,
                            data_2,
                            category_1,
                            category_2,
                            output) {
        tmp_1 <- extract_snps(all_snps = data_1,
                              snps_of_interest = data_2,
                              category_of_snp  = category_1)
        tmp_2 <- extract_snps(all_snps = data_1,
                              snps_of_interest = data_2,
                              category_of_snp  = category_2)
        print(chisq.test(rbind(tmp_1,
                               tmp_2)))
        output <- chisq.test(rbind(tmp_1,
                               tmp_2))
        print(output$expected)
        print(output$observed)
}
```

5. Call function for:
a) Intronic SNPs vs coding regions.
b) Intergenic SNPs vs coding regions.
c) 5kb up/downstream SNPs vs coding regions.

```{r, message = FALSE}
run_chiseq_test(data_1 = data,
                data_2 = nsl_top_one_percent_df,
                category_1 = "Coding",
                category_2 = "Intron",
                output = test)

run_chiseq_test(data_1 = data,
                data_2 = nsl_top_one_percent_df,
                category_1 = "Coding",
                category_2 = "Intergenic")
run_chiseq_test(data_1 = data,
                data_2 = nsl_top_one_percent_df,
                category_1 = "Coding",
                category_2 = "stream")
```

6. Run lintr:

```{r, message = FALSE}
require(lintr)
lintr::lint(file = "run_chisquare_analysis.Rmd")
```
