---
title: "Bombus population genomics"
output: run_haplotype_correlation.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

##Â Introduction:  
This script takes a VCF file as input (VCF generated by Freebayes) and using gdsfmt generates a genotype matrix.  
Using this genotypic matrix, correlations in genotypes between two genes are estimated.

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("gdsfmt",
               "SNPRelate",
               "ggplot2",
               "ggpubr",
               "lintr",
               "reshape2")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE )
    }
}
## Create output directory:
dir.create("results")
```

Load data and perform PCA:

```{r}
## Plot PCA of larger number of samples (n=46):
## Assign the input name:  
haplotype_1_vcf <- "data/haplotype_1.bed.recode.vcf"
haplotype_2_vcf <- "data/haplotype_2.bed.recode.vcf"

## Convert VCF to genome data structure (gds) file format
## gds is a container for storing annotation data and SNP genotypes.
## In this format, each byte encodes up to four SNP genotypes thereby reducing file size and access time.
## The GDS format supports data blocking so that only the subset of data that is being processed needs
## to reside in memory. GDS formatted data is also designed for efficient data access to large datasets.

## Use the name of input to designate the name for the GDS file format.
haplotype_1_vcf_gds <- gsub(".vcf",
                            ".gds",
                            haplotype_1_vcf)

haplotype_2_vcf_gds <- gsub(".vcf",
                            ".gds",
                            haplotype_2_vcf)

## Reformat VCF into GDS format
snpgdsVCF2GDS(haplotype_1_vcf,
              haplotype_1_vcf_gds,
              method = "biallelic.only")

snpgdsVCF2GDS(haplotype_2_vcf,
              haplotype_2_vcf_gds,
              method = "biallelic.only")

## Extract genotype information:
haplotype_1_vcf_gds_genofile <- snpgdsOpen(haplotype_1_vcf_gds)
haplotype_2_vcf_gds_genofile <- snpgdsOpen(haplotype_2_vcf_gds)

## Extract genotype information and generate genotypic matrix:  
haplotype_1_vcf_gds_genofile_matrix <- read.gdsn(index.gdsn(haplotype_1_vcf_gds_genofile,
                                                      "genotype"))

haplotype_2_vcf_gds_genofile_matrix <- read.gdsn(index.gdsn(haplotype_2_vcf_gds_genofile,
                                                      "genotype"))
                                                      
## For performing PCA, it is recommended to perform LD pruning:
#snpset <- snpgdsLDpruning(input_vcf_gds_genofile,
#                          ld.threshold = 0.1,
#                          autosome.only = FALSE)
estimate <- vector()
p_values <- vector()
for (sample in 1:nrow(haplotype_1_vcf_gds_genofile_matrix)){
        test <- cor.test(haplotype_1_vcf_gds_genofile_matrix[, sample],
            haplotype_2_vcf_gds_genofile_matrix[, sample])
        estimate[sample] <- test$estimate
        p_values[sample] <- test$p.value
}
```

Run lintr:

```{r, message = FALSE}
lintr::lint(file = "./run_haplotype_correlation.Rmd")
```
