--- 
title: "Bombus pesticide exposure: gene expression in head"
output: gene_level_3treatments.html
authors:  Joe Colgan (joscolgan), Yannick Wurm http://wurmlab.com
--- 

# Introduction
This script is for differential gene expression analysis of transcript quantification data.
Input files should be abundance files (.h5, .tsv, .json) generated by [Kallisto](https://pachterlab.github.io/kallisto/manual) 
for each sample and .txt file of genome transcripts 
and their corresponding gene ids (LOC numbers)/.

This script will import abundance files from Kallisto and use the R package [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq.html) on count data to test for 
differential expression based on a model using negative binomial distribution.
This script produces a results table of genes that are significantly differentially expressed. 
This script uses the Wald test in DESeq2 to test for differential expression between treatments.
This can then be used for further analysis. 


```{r}
# Load libraries; install from scratch if needed
libraries <- c("tximport",
               "readr",
               "DESeq2",
               "ggplot2",
               "gplots",
               "rhdf5",
               "plyr",
               "ggfortify",
               "biomaRt",
               "vsn",
               "RColorBrewer",
               "ReportingTools")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE )
    }
}
dir.create("results")
```


## Step One: Input results from kallisto

We use raw, non-normalised estimated counts.
```{r} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$sample_batch_info <- "data/sample_information.tmp"
paths$kallisto_output   <- "input/"

# Set relative paths:
paths$kallisto_files_relative <- grep(x = list.files(paths$kallisto_output,
                                                     recursive = TRUE),
                                      pattern = ".h5",
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

# Automatically extract file names from kallisto output for colnames
names(paths$kallisto_files) <- gsub(paths$kallisto_files_relative,
                                    pattern = "/.*",
                                    replacement = "")

for (filenumber in 1:length(paths$kallisto_files)) {
  current_name <- names(paths$kallisto_files)[filenumber]
  current_file <- paths$kallisto_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
```


## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r}
## Extract sample names and put into df for tximport
samples     <- data.frame(treatment = names(paths$kallisto_files))

## Read in sample information for room and batch:
samples_information <- read.table(file = paths$sample_batch_info,
                                  header = FALSE,
                                  col.names = c("sample_name",
                                                "sex_caste",
                                                "tissue"),
                                  row.names = 1)

# sanity check
if (FALSE == all(samples$treatment %in% rownames(samples_information))) {
    kill("we have a problem - sample names from files and
         from batch descriptor file  dont match up")
}
samples_information
```
## Step Three: Import kallisto quantification files using Tximport

We used kallisto to obtain estimated counts for each transcript in each condition. 
We load these raw counts into R for DEseq using tximport. 
Because we do a gene-level analysis first, transcript abundances are summarised per gene during import.

```{r}
# Read in file corresponding to transcripts to gene ids
transcript_to_gene <- read.table(file = "./data/rna_and_corresponding_gene_ids.txt",
                                 col.names = c("transcript",
                                               "locus"))

# Use tximport on kallisto files 
## Counts are estimated counts from kallisto:
txi_counts <- tximport(paths$kallisto_files,
                       type    = "kallisto",
                       tx2gene = transcript_to_gene,
                       countsFromAbundance = "lengthScaledTPM")

# Save object for use in other scripts
dir.create(path = "results/count_estimates/")
save(txi_counts, file = "results/count_estimates/txi_count_estimates.Rdata")

## Generate plot:
## Subset gene of interest:
genes_of_interest <- "LOC105666162"
stage_txi_counts_df <- txi_counts$counts
genes_of_interest_df <- subset(stage_txi_counts_df,
                               row.names(stage_txi_counts_df)
                               %in% genes_of_interest)

## Read in sample information:
sample_info <- read.table(file = "./data/sample_information.tmp",
                          header = FALSE)

genes_of_interest_df_tmp <- as.data.frame(t(genes_of_interest_df))
genes_of_interest_df_tmp$sex <- sample_info$V2
genes_of_interest_df_tmp$stage <- sample_info$V3
genes_of_interest_df_tmp$samples <- row.names(genes_of_interest_df_tmp)
genes_of_interest_df_tmp$tissue <- paste(genes_of_interest_df_tmp$sex,
                                         "_",
                                         genes_of_interest_df_tmp$stage,
                                         sep = "")

genes_of_interest_df_tmp$tissue <- gsub(pattern = "_",
                                        replacement = " ",
                                        genes_of_interest_df_tmp$tissue)

## Melt:
require(reshape2)
genes_of_interest_melt <- melt(genes_of_interest_df_tmp)

genes_of_interest_melt$sample_id <- paste(
        genes_of_interest_melt$tissue,
      rownames(genes_of_interest_melt),
      sep = " ")

levels(genes_of_interest_melt$sample_id) <-
       genes_of_interest_melt$sample_id

genes_of_interest_melt$sample_id <- factor(genes_of_interest_melt$sample_id,
       levels(genes_of_interest_melt$sample_id))

## Extract list of samples based on developmental stage:
adult_samples <- grep("adult",
                      genes_of_interest_melt$sample_id)
pupae_samples <- grep("pupae",
                      genes_of_interest_melt$sample_id)
larvae_samples <- grep("larvae",
                      genes_of_interest_melt$sample_id)
combined_samples <- c(larvae_samples,
                      pupae_samples,
                      adult_samples)
other_samples <- subset(genes_of_interest_melt,
       !(rownames(genes_of_interest_melt) %in% combined_samples))

## Reorder dataframe to plot based on developmental stage:
new_order_df <- rbind(subset(genes_of_interest_melt,
       rownames(genes_of_interest_melt) %in% larvae_samples),
subset(genes_of_interest_melt,
       rownames(genes_of_interest_melt) %in% pupae_samples),
subset(genes_of_interest_melt,
       rownames(genes_of_interest_melt) %in% adult_samples),
subset(genes_of_interest_melt,
       rownames(genes_of_interest_melt) %in% other_samples))

## Reorder the samples for plotting:
new_order_df$sample_id <- factor(new_order_df$sample_id,
       levels = as.character(new_order_df$sample_id))

replacement_order <- c(as.character(new_order_df$sample_id[1:14]),
       as.character(new_order_df$sample_id[18:20]),
       as.character(new_order_df$sample_id[15:17]),
       as.character(new_order_df$sample_id[21:27]))

## Update worker status for plotting:
replacement_order <- gsub("rep_",
                               "",
                               replacement_order)
replacement_order <- gsub("undet_",
                               "",
                               replacement_order)

new_order_df$sex <- gsub("rep_",
                         "",
                         new_order_df$sex)
new_order_df$sex <- gsub("undet_",
                         "",
                         new_order_df$sex)

new_order_df$sample_id <- factor(new_order_df$sample_id,
       levels = unique(as.character(replacement_order)))

colnames(new_order_df)[1] <- "caste"

## Generate plot:
life_wolbachia_stages_plot <- ggplot(data = new_order_df,
                           aes(x = sample_id,
                               y = value)) +
        geom_point(aes(fill = caste),
                   colour = "black",
                   pch = 21,
                   alpha = 0.8,
                   size = 5) +
        ylim(0, 3000) +
        geom_vline(xintercept = 7.5,
                   linetype = "dashed") +
        geom_vline(xintercept = 14.5,
                   linetype = "dashed") +
        ylab("Length-scaled TPM") +
        xlab("Samples") +
        scale_fill_manual(values = c("light blue",
                                     "orange",
                                     "yellow",
                                     "brown")) +
        theme_bw() +
        theme(axis.title = element_text(face = "bold"),
              legend.text = element_text(size = 10,
                                         face = "plain"),
              legend.title = element_text(size = 10,
                                          face = "bold"),
              legend.justification = c(1.1, 1.1),
              legend.position = c(1, 1),
              axis.title.x = element_blank(),
                axis.text.x = element_text(angle = 45,
                                           hjust = 1))

## Add annotation to plot:
life_wolbachia_stages_plot <- life_wolbachia_stages_plot +
        annotate("text",
                 x = 4,
                 y = 2800,
                 label = "Larval stage",
                 fontface = 2) +
        annotate("text",
                 x = 11,
                 y = 2800,
                 label = "Pupal stage",
                 fontface = 2) +
        annotate("text",
                 x = 18,
                 y = 2800,
                 label = "Adult stage",
                 fontface = 2)

## Save plot:
save(life_wolbachia_stages_plot,
     file = "./results/wolbachia_gene_expression.Rdata")
```

Run lintr:

```{r, message = FALSE}
lintr::lint(file = "./gene_level_wolbachia_expression_plus_plot.Rmd")
```
