

```{r, message = FALSE}
## Load libraries:
libraries <- c("ggplot2",
               "SNPRelate",
               "reshape2",
               "ggpubr",
               "stringr",
               "gggenes")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}
```

Read in nucleotide diversity estimates for each of the four global populations:

```{r, message = FALSE}
nuc_div_africa <- read.table(file = "./africa/apis_all_chrom_10kb.txt",
                             header = TRUE)
nuc_div_asia <- read.table(file = "./asia/apis_all_chrom_10kb.txt",
                             header = TRUE)
nuc_div_weurope <- read.table(file = "./western_europe/apis_all_chrom_10kb.txt",
                             header = TRUE)
nuc_div_eeurope <- read.table(file = "./eastern_europe/apis_all_chrom_10kb.txt",
                             header = TRUE)
```

Subset nucleotide diversity and combine into a new dataframe:

```{r, message = FALSE}
## Generate combined df:
combined_pop_df <- data.frame()
combined_pop_df <- nuc_div_africa[, c(1:3, 7)]
combined_pop_df$africa_nd <- nuc_div_africa$nuc_diversity      
combined_pop_df$asia_nd <- nuc_div_asia$nuc_diversity      
combined_pop_df$weurope_nd <- nuc_div_weurope$nuc_diversity      
combined_pop_df$eeurope_nd <- nuc_div_eeurope$nuc_diversity      
```

```{r, message=FALSE}
## Assign the coordinates of the gene to variables plus 100kb 
## up and downstream of the gene:
chrom <- "NC_037638.1"
region_one_start <- 20537332 - 100000
region_one_end <- 20625795 + 100000

region_two_start <- 24977996 - 100000 
region_two_end <- 25136877 + 100000

large_genomic_region_start <- 20537332
large_genomic_region_end <- 25136877
```

Subset genomic region to plot nucleotide diversity:

```{r, message = FALSE}
## Subset region:
nuc_div_subset_one <- subset(combined_pop_df,
                         chrom == "NC_037638.1" &
                         start > region_one_start &
                         end < region_one_end)

nuc_div_subset_two <- subset(combined_pop_df,
                         chrom == "NC_037638.1" &
                         start > region_two_start &
                         end < region_two_end)

nuc_div_subset_large <- subset(combined_pop_df,
                         chrom == "NC_037638.1" &
                         midpoint > large_genomic_region_start &
                         midpoint < large_genomic_region_end)

## Calculate mean nucleotide diversity:
nucdiv_median <- median(combined_pop_df$africa_nd)
nuc_div_subset_one$chrom <- NULL
nuc_div_subset_one$start <- NULL
nuc_div_subset_one$end <- NULL

## Melt dataframe:
nuc_div_subset_one_melted <- melt(nuc_div_subset_one,
                                  id.vars = "midpoint")

nuc_div_subset_two$chrom <- NULL
nuc_div_subset_two$start <- NULL
nuc_div_subset_two$end <- NULL

## Melt dataframe:
nuc_div_subset_two_melted <- melt(nuc_div_subset_two,
                                  id.vars = "midpoint")

## Melt large region:
nuc_div_subset_large$chrom <- NULL
nuc_div_subset_large_melted <- melt(nuc_div_subset_large,
                                    id.vars = "midpoint")
```


```{r, message = FALSE}
## Generate plot:
region_one_plot <- ggplot(nuc_div_subset_one_melted,
                       aes(x = midpoint,
                           y = value,
                           colour = variable)) +
        geom_point(colour = "black",
                   size = 0,
                   alpha = 0.75) +
        geom_line() +
        xlab("Genomic coordinates (bp)") +
        ylab("Nucleotide diversity") +
        geom_hline(yintercept = c(nucdiv_median),
                   linetype = "dashed",
                   colour = "orange") +
        geom_rect(aes(xmin = 20537332,
                      xmax = 20625795,
                      ymin = -Inf,
                      ymax = Inf),
                  fill = "yellow",
                  alpha = 0.008) +
        scale_x_continuous(labels = scales::comma,
                           limits = c(region_one_start,
                                      region_one_end),
                           position = "top") +
        scale_color_manual(values = c("blue",
                                      "orange",
                                      "black",
                                      "purple")) +
        theme_bw() +
        theme(axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title = element_text(size = 15,
                                        face = "bold"),
              axis.title.x = element_blank())
```

```{r, message = FALSE}
## Generate plot:
region_two_plot <- ggplot(nuc_div_subset_two_melted,
                       aes(x = midpoint,
                           y = value,
                           colour = variable)) +
        geom_point(colour = "black",
                   size = 0,
                   alpha = 0.75) +
        geom_line() +
        xlab("Genomic coordinates (bp)") +
        ylab("Nucleotide diversity") +
        geom_hline(yintercept = c(nucdiv_median),
                   linetype = "dashed",
                   colour = "orange") +
        geom_rect(aes(xmin = 24977996,
                      xmax = 25136877,
                      ymin = -Inf,
                      ymax = Inf),
                  fill = "yellow",
                  alpha = 0.008) +
        scale_x_continuous(labels = scales::comma,
                           limits = c(region_two_start,
                                      region_two_end),
                           position = "top") +
        scale_color_manual(values = c("blue",
                                      "orange",
                                      "black",
                                      "purple")) +
        theme_bw() +
        theme(axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title = element_text(size = 15,
                                        face = "bold"),
              axis.title.x = element_blank())
```

```{r, message = FALSE}
## Generate plot:
region_large_plot <- ggplot(nuc_div_subset_large_melted,
                       aes(x = midpoint,
                           y = value,
                           colour = variable)) +
        geom_point(colour = "black",
                   size = 0,
                   alpha = 0.75) +
        geom_line() +
        xlab("Genomic coordinates (bp)") +
        ylab("Nucleotide diversity") +
        geom_hline(yintercept = c(nucdiv_median),
                   linetype = "dashed",
                   colour = "orange") +
        geom_rect(aes(xmin = 24977996,
                      xmax = 25136877,
                      ymin = -Inf,
                      ymax = Inf),
                  fill = "yellow",
                  alpha = 0.008) +
        scale_x_continuous(labels = scales::comma,
                           limits = c(large_genomic_region_start,
                                      large_genomic_region_end),
                           position = "top") +
        scale_color_manual(values = c("blue",
                                      "orange",
                                      "black",
                                      "purple")) +
        theme_bw() +
        theme(axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title = element_text(size = 15,
                                        face = "bold"),
              axis.title.x = element_blank())
```

Plot together:

```{r, message = FALSE}
ggarrange(region_one_plot,
          region_two_plot,
          ncol = 2,
          nrow = 1,
          labels = c("A", "B"),
          align = "hv",
          common.legend = TRUE)

## Save to file:
ggsave(file = "./combined_pop_plots.png",
       height = 5,
       width = 8)
```