---
title: "Bombus population genomics"
output: plot_transcription_factor_sweeps.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:
This script takes four tab-delimited files as input:
1) A file containing absolute nsl scores assigned to each SNP in the British population.
2) A file containing absolute iHS scores assigned to each SNP in the British population. Both the nsl and iHS scores were calculated by [selscan](https://github.com/szpiech/selscan).
3) A file containing population-scaled recombination rate as calculaed by [ldhelmet](https://sourceforge.net/projects/ldhelmet/).
4) A file containing nucleotide diversity estimates for sliding windows of 10kb as calculated by [popgenome](https://cran.r-project.org/web/packages/PopGenome/index.html).
The input files are used to generate individual scatterplots using [ggplot2](https://ggplot2.tidyverse.org/).
The script also loads arrow-plots and heatmaps generated as output by other scripts.
The scripts generates a combined multiple graph using [ggpubr](https://rpkgs.datanovia.com/ggpubr/index.html), which is output to file.
Code style is checked using [lintr](https://cran.r-project.org/web/packages/lintr/index.html).

```{r, message = FALSE}
## Load libraries:
libraries <- c("ggplot2",
               "SNPRelate",
               "reshape2",
               "ggpubr",
               "stringr",
               "gggenes")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}
```

This script plots nsl score, ihs score, population-scaled recombination rate, nucleotide diversity and Tajima's D estimates.

```{r, message = FALSE}
## Read in data:
nsl_data <- read.table(file = "./input/nsl_output/NC_015765.1.nsl.out.100bins.norm.abs.bed.tmp",
                       col.names = c("chrom",
                                     "position",
                                     "position",
                                     "nsl_score"))
ihs_data <- read.table(file = "./input/ihs_output/NC_015765.1.ihs.out.100bins.norm.abs.bed.tmp",
                       col.names = c("chrom",
                                     "position",
                                     "position",
                                     "ihs_score"))
recomb_data <- read.table(file = "./input/recombination_rate/NC_015765.1.input_for_genetic_map.txt",
                          header = FALSE,
                          col.names = c("chrom",
                                        "start_pos",
                                        "rho",
                                        "physical_distance"))
## Both nucleotide diversity and Tajima D are in the same input file:
nuc_div_data <- read.table(file = "./input/nucleotide_diversity_output/bter_all_chrom_10kb.txt",
                           header = TRUE)
```

The gene of interest is an uncharacterised gene (LOC100651242), which codes for a protein of unknown function. The gene is located on chromosome 11:
NC_015764.1 (11050477...11084406)

```{r, message=FALSE}
## Assign the coordinates of the gene to variables plus 100kb 
## up and downstream of the gene:
chrom <- "NC_015765.1"
central_position <- 771541
start <- central_position - 100000
end <- central_position + 80000
end_remove <- 4052
```

Subset the regions of interest:

```{r, message=FALSE}
## Subset nsl values for plotting:
nsl_input_subset <- subset(nsl_data,
                            position > start &
                            position < 1000000)

## subset sites above 2.56 to highlight on plot:
nsl_snps_highlight <- subset(nsl_input_subset,
                            nsl_score > 2.56)

## Subset ihs values for plotting:
ihs_input_subset <- subset(ihs_data,
                            position > start &
                            position < 1000000)

ihs_snps_highlight <- subset(ihs_input_subset,
                            ihs_score > 2.89)
```

Generate plot of EHH:

```{r, message = FALSE}
region_to_remove_size <- 51242
region_to_correct     <- nsl_input_subset[which(rownames(nsl_input_subset) > end_remove), ]

## From start, end and midpoint 
region_to_correct$position    <- as.numeric(unlist(region_to_correct$position)) - region_to_remove_size
## Remove shitebag sites:
## Update original dataframe with corrected values:
nsl_input_subset[match(rownames(region_to_correct), rownames(nsl_input_subset)), ] <- region_to_correct

nsl_plot <- ggplot(nsl_input_subset,
                   aes(x = position,
                       y = nsl_score)) +
                xlab("Genomic coordinates (bp)") +
                ylab("|nsl| value") +
         geom_rect(aes(xmin = 725000,
                              xmax = 825000,
                              ymin = -Inf,
                              ymax = Inf),
                          fill = "light blue",
                          alpha = 0.008) +
        geom_hline(yintercept = c(2.56),
                   linetype = "dashed",
                   colour = "orange") +
        geom_vline(xintercept = c(725000,
                                  825000),
                   linetype = "dashed",
                   colour = "blue") +
        scale_x_continuous(labels = scales::comma,
                           limits = c(700000, 850000)) +
                geom_point(colour = "black",
                           size = 3,
                           alpha = 0.5) +
                geom_point(data = nsl_snps_highlight,
                           aes(x = position,
                               y = nsl_score),
                           colour = "blue",
                           pch = 21,
                           size = 3,
                           alpha = 0.75) +
                theme_bw() +
        theme(axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title.y = element_text(size = 15,
                                          face = "bold"),
              axis.title.x = element_blank(),
              axis.text.x = element_blank(),
              axis.ticks.x = element_blank())
```

Generate the ihs plot:

```{r, message = FALSE}
region_to_remove_size <- 51242
region_to_correct     <- ihs_input_subset[which(rownames(ihs_input_subset) > end_remove), ]

## From start, end and midpoint 
region_to_correct$position    <- as.numeric(unlist(region_to_correct$position)) - region_to_remove_size
## Remove shitebag sites:
## Update original dataframe with corrected values:
ihs_input_subset[match(rownames(region_to_correct), rownames(ihs_input_subset)), ] <- region_to_correct

ihs_plot <- ggplot(ihs_input_subset,
                   aes(x = position,
                       y = ihs_score)) +
                xlab("Genomic coordinates (bp)") +
                ylab("|ihs| value") +
                geom_rect(aes(xmin = 764617,
                              xmax = 879060,
                              ymin = -Inf,
                              ymax = Inf),
                          fill = "light blue",
                          alpha = 0.008) +
        geom_hline(yintercept = c(2.89),
                   linetype = "dashed",
                   colour = "orange") +
        scale_x_continuous(labels = scales::comma,
                           limits = c(671541, 851541)) +
                geom_point(colour = "black",
                           size = 3,
                           alpha = 0.5) +
                geom_point(data = ihs_snps_highlight,
                           aes(x = position,
                               y = ihs_score),
                           colour = "blue",
                           size = 3,
                           alpha = 0.5) +
                theme_bw()
```

Subset genomic region to plot nucleotide diversity:

```{r, message = FALSE}
## Calculate median nucleotide diversity:
nucdiv_median <- median(nuc_div_data$nuc_diversity)

## Genomic region to correct for plotting:
region_to_correct     <- subset(nuc_div_data,
                                chrom == "NC_015765.1" &
                                start > 860001)

## Genomic regions of NAs to remove:
region_to_remove     <- subset(nuc_div_data,
                                chrom == "NC_015765.1" &
                                start > 810001 &
                                end < 860001)

## Size of genomic region to correct for:
region_to_remove_size <-  tail(region_to_remove$end, n = 1) -
                          head(region_to_remove$start, n = 1)

## From start, end and midpoint 
region_to_correct$start    <- as.numeric(unlist(region_to_correct$start)) - region_to_remove_size
region_to_correct$end      <- as.numeric(unlist(region_to_correct$end)) - region_to_remove_size
region_to_correct$midpoint <- as.numeric(unlist(region_to_correct$midpoint)) - region_to_remove_size

## Update original dataframe with corrected values:
nuc_div_data[match(rownames(region_to_correct), rownames(nuc_div_data)), ] <- region_to_correct
nuc_div_data <- nuc_div_data[ !(rownames(nuc_div_data) %in% rownames(region_to_remove)), ]

## Subset corrected genomic regions to plot:
nuc_div_subset <- subset(nuc_div_data,
                         chrom == "NC_015765.1" &
                         start > 664617 &
                         end < 1000000)

options(scipen = 999)

## Plot:
nuc_div_plot <- ggplot(nuc_div_subset,
                       aes(x = midpoint,
                           y = nuc_diversity)) +
        xlab("Genomic position (bp)") +
        ylab("Nucleotide diversity") +
        geom_vline(xintercept = c(725000,
                                  825000),
                   linetype = "dashed",
                   colour = "blue") +
        geom_hline(yintercept = c(nucdiv_median),
                   linetype = "dashed",
                   colour = "orange") +
        geom_rect(aes(xmin = 725000,
                      xmax = 825000,
                      ymin = -Inf,
                      ymax = Inf),
                  fill = "light blue",
                  alpha = 0.008) +
        geom_point(colour = "black",
                   size = 4,
                   alpha = 0.75) +
        scale_x_continuous(labels = scales::comma,
                           limits = c(700000, 850000)) +
        theme_bw() +
        theme(axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title = element_text(size = 15,
                                        face = "bold"))
```

Load SNP matrix for combined plot:

```{r, message = FALSE}
load(file = "./results/plots/transcription_factor_snp_plot.Rdata")
load(file = "./results/plots/tf_arrow_plot.Rdata")
```

Generate combined plots:

```{r, message=FALSE}
## Generate a combined plot:
ggarrange(arrow_plot,
          nuc_div_plot,
          nsl_plot,
          heatmap_plot,
          nrow = 4,
          ncol = 1,
          align = "hv",
          heights = c(0.7, 2, 1.8, 3),
          labels = c("a",
                     "b",
                     "c",
                     "d"))

## Save:
ggsave(file = "results/transcription_factor_combined_plot.png",
       width = 10,
       height = 12)
```

Run lintr:

```{r, message = FALSE}
## run lintr:
lintr::lint(file = "./plot_transcription_factor_sweeps.Rmd")

```

