---
title: "Bombus population genomics"
output: make_gggenes_plot.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:  
This script takes a tab-delimited file as input containing the following columns:
1) Chromosome name
2) Start position of chromosome
3) End position of chromosome
4) Gene name
5) Strand
6) Status - if you want to highlight a gene of interest

N.B. I tested out the code using the example_data that comes with the package and everything worked. Created my own dataframe using the same column headers in R but there was an issue with the plots making all genes
the same size in the plot. Unsure of what the problem is. Easiest fix
was to output the example data file and manually change values for plots. 
No issues since. 

```{r}
## Load libraries:
library(ggplot2)
library(gggenes)
library(tidyverse)
library(ggpubr)
library(lintr)
```

Just plot Bombus - highlighting regions of similarity to Wolbachia and other insects

```{r, message = FALSE}
## Read in file for plotting:
data <- read.table(file = "./input/input_for_plot_domains.txt",
                   header = TRUE)
colnames(data)

protein_domain <- ggplot(data, aes(xmin = start,
                         xmax = end,
                         y = protein)) +
        facet_wrap(~ protein,
                   scales = "free",
                   ncol = 1) +
        geom_gene_arrow(arrowhead_height = unit(3, "mm"),
                        arrowhead_width = unit(1, "mm")) +
        geom_subgene_arrow(data = data,
                           aes(xmin = start,
                               xmax = end,
                               y = protein,
                               fill = domain,
                               xsubmin = start_domain,
                               xsubmax = end_domain),
                           color = "black",
                           alpha = .7) +
        theme_genes() +
        theme(axis.ticks = element_blank(),
              axis.title.y = element_blank(),
              axis.text = element_blank(),
              axis.ticks.x = element_blank(),
              axis.line.x = element_blank(),
              #legend.title = element_text(face = "bold"),
              legend.position = "none")

## Update colours for plotting:
protein_domain <- protein_domain +
        scale_fill_manual(name = "Functional domains",
                          values = c("yellow",
                                     "blue")) +
        theme(plot.margin = grid::unit(c(0, 0, 0, 0), "mm"))
```

Plot piecharts for four sections of the predicted protein:

First, create a blank background for plotting:

```{r, message = FALSE}
## Create a blank theme for the background of each plot:
blank_theme <- theme_minimal() +
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid = element_blank(),
  axis.ticks = element_blank(),
  plot.title = element_text(size = 8,
                            face = "bold")
  )
```

Read in first subsection for plotting:

```{r, message = FALSE}
## Read in files containing:
## Group information and number of protein matches per group:
## Read in files for each subsection:
sequence_1_info <- read.table(file = "./input/group_counts/sequence_one_count_modified.txt",
                              header = TRUE)

## The second column ('group') contains words joined by underscores.
## For plotting, replace underscores with spaces.
sequence_1_info$group <- gsub(pattern = "_",
                              replacement = " ",
                              sequence_1_info$group)

## Update levels for plotting:
sequence_1_info$group <- factor(sequence_1_info$group,
                                   levels = unique(sequence_1_info$group))

## Calculate the position for label of count for each group in the chart:
sequence_1_info <- sequence_1_info %>%
        arrange(desc(group)) %>%
        mutate(lab.ypos = cumsum(count) - 0.5 * count)

## Generate plot:
sequence_1_piechart <- ggplot(sequence_1_info,
                              aes(x = "",
                                  y = count,
                                  fill = group)) +
        geom_bar(width = 1,
                 stat = "identity",
                 color = "white") +
        coord_polar("y", start = 0) +
        geom_text(data = subset(sequence_1_info, count > 0),
                  aes(y = lab.ypos,
                      label = count),
                  size = 3,
                  color = "white") +
        scale_fill_manual(name = "Taxonomic\ngroup",
                          values = c("steelblue1",
                                    "light blue",
                                     "blue",
                                     "navy",
                                     "orange",
                                     "brown",
                                     "grey",
                                     "black")) +
        theme_void() +
        theme(axis.title = element_blank(),
              legend.position = "none",
              legend.title = element_text(face = "bold")) +
        theme(plot.margin = grid::unit(c(0, 0, 0, 0), "mm"))

## Reformat the legend positions:
sequence_1_piechart <- sequence_1_piechart +
        guides(fill = guide_legend(nrow = 3,
                                   ncol = 3,
                                   byrow = TRUE))
```

Read in second subsection for plotting:

```{r, message = FALSE}
## read in sequence 2:
sequence_2_info <- read.table(file = "./input/group_counts/sequence_two_count_modified.txt",
                              header = TRUE)

## The second column ('group') contains words joined by underscores.
## For plotting, replace underscores with spaces.
sequence_2_info$group <- gsub(pattern = "_",
                              replacement = " ",
                              sequence_2_info$group)

## Update levels for plotting:
sequence_2_info$group <- factor(sequence_2_info$group,
                                   levels = unique(sequence_2_info$group))

## Calculate the position for label of count for each group in the chart:
sequence_2_info <- sequence_2_info %>%
        arrange(desc(group)) %>%
        mutate(lab.ypos = cumsum(count) - 0.5 * count)

## Generate plot:
sequence_2_piechart <- ggplot(sequence_2_info,
                              aes(x = "",
                                  y = count,
                                  fill = group)) +
        geom_bar(width = 1,
                 stat = "identity",
                 color = "white") +
        coord_polar("y", start = 0) +
        geom_text(data = subset(sequence_2_info, count > 0),
                  aes(y = lab.ypos,
                      label = count),
                  size = 3,
                  color = "white") +
        scale_fill_manual(name = "Taxonomic\ngroup",
                          values = c("steelblue1",
                                    "light blue",
                                     "blue",
                                     "navy",
                                     "orange",
                                     "brown",
                                     "grey",
                                     "black")) +
        theme_void() +
        theme(axis.title = element_blank(),
              legend.position = "none",
              legend.title = element_text(face = "bold")) +
        theme(plot.margin = grid::unit(c(0, 0, 0, 0), "mm"))
```

Read in third subsection for plotting:

```{r, message = FALSE}
## read in sequence 3:
sequence_3_info <- read.table(file = "./input/group_counts/sequence_three_count_modified.txt",
                              header = TRUE)

## The second column ('group') contains words joined by underscores.
## For plotting, replace underscores with spaces.
sequence_3_info$group <- gsub(pattern = "_",
                              replacement = " ",
                              sequence_3_info$group)

## Update levels for plotting:
sequence_3_info$group <- factor(sequence_3_info$group,
                                   levels = unique(sequence_3_info$group))

## Calculate the position for label of count for each group in the chart:
sequence_3_info <- sequence_3_info %>%
        arrange(desc(group)) %>%
        mutate(lab.ypos = cumsum(count) - 0.5 * count)

## Generate plot:
sequence_3_piechart <- ggplot(sequence_3_info,
                              aes(x = "",
                                  y = count,
                                  fill = group)) +
        geom_bar(width = 1,
                 stat = "identity",
                 color = "white") +
        coord_polar("y", start = 0) +
        geom_text(data = subset(sequence_3_info, count > 0),
                  aes(y = lab.ypos,
                      label = count),
                  size = 3,
                  color = "white") +
        scale_fill_manual(name = "Taxonomic\ngroup",
                          values = c("steelblue1",
                                    "light blue",
                                     "blue",
                                     "navy",
                                     "orange",
                                     "brown",
                                     "grey",
                                     "black")) +
        theme_void() +
        theme(axis.title = element_blank(),
              legend.position = "none",
              legend.title = element_text(face = "bold")) +
        theme(plot.margin = grid::unit(c(0, 0, 0, 0), "mm"))
```

Read in fourth and final subsection for plotting:

```{r, message = FALSE}
## read in sequence 4:
sequence_4_info <- read.table(file = "./input/group_counts/sequence_four_count_modified.txt",
                              header = TRUE)

## The second column ('group') contains words joined by underscores.
## For plotting, replace underscores with spaces.
sequence_4_info$group <- gsub(pattern = "_",
                              replacement = " ",
                              sequence_4_info$group)

## Update levels for plotting:
sequence_4_info$group <- factor(sequence_4_info$group,
                                   levels = unique(sequence_4_info$group))

## Calculate the position for label of count for each group in the chart:
sequence_4_info <- sequence_4_info %>%
        arrange(desc(group)) %>%
        mutate(lab.ypos = cumsum(count) - 0.5 * count)

## Generate plot:
sequence_4_piechart <- ggplot(sequence_4_info,
                              aes(x = "",
                                  y = count,
                                  fill = group)) +
        geom_bar(width = 1,
                 stat = "identity",
                 color = "white") +
        coord_polar("y", start = 0) +
        geom_text(data = subset(sequence_4_info, count > 0),
                  aes(y = lab.ypos,
                      label = count),
                      size = 3,
                  color = "white") +
        scale_fill_manual(name = "Taxonomic\ngroup",
                          values = c("steelblue1",
                                    "light blue",
                                     "blue",
                                     "navy",
                                     "orange",
                                     "brown",
                                     "grey",
                                     "black")) +
        theme_void() +
        theme(axis.title = element_blank(),
              legend.position = "none",
              legend.title = element_text(face = "bold")) +
        theme(plot.margin = grid::unit(c(0, 0, 0, 0), "mm"))
```

Combine plots:

```{r, message = FALSE}
## Generate a combined plot for all four piecharts:
combined_pies <- ggarrange(sequence_1_piechart,
                           sequence_2_piechart,
                           sequence_3_piechart,
                           sequence_4_piechart,
                           nrow = 1,
                           ncol = 4,
                           align = "hv",
                           legend = "bottom",
                           common.legend = TRUE)

## Combine the arrow plot and piecharts for a combined plot:
combined_fig <- ggarrange(protein_domain,
                          combined_pies,
                          ncol = 1,
                          nrow = 2,
                          heights = c(0.5,
                                      2),
                          align = "hv")

## Output plot to file:
ggsave(file = "./results/protein_domain.png",
       height = 6,
       width = 10)

## Save everything as R object:
save.image(file = "./results/protein_domain_plus_pies.RData")
```

Run lintr:

```{r, message = FALSE}
lintr::lint(file = "./make_gggenes_plot.Rmd")
```