---
title: "Bombus population genomics"
output: plot_phylogenetic_tree_modified.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

The purpose of this script is to take a newick-aligned dataset of protein sequences and plot the phylogeny. 

## Load libraries:

```{r, message = FALSE}
## Load libraries:
libraries <- c("tidyverse",
               "ggtree",
               "rphylopic",
               "ggplot2",
               "EBImage")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}
## Create output directory:
dir.create(path = "./plot")
```

Read in input data in newick format.
Generate plot and save to file.

```{r, message = FALSE}
## Read newick alignment file in and store as variable:
tree <- read.tree("./RAxML_bipartitions.putative_homologues_refseq_names_edited.afa_n100.tree")

## A preliminary tree to check internal nodes
ggtree(tree) + geom_text2(aes(subset = !isTip,
                              label = label),
                          hjust = -.3) + geom_tiplab()


## Save:
ggsave(file = "./results/figures/phylo_plot_n100_bs.png",
       height = 15,
       width = 20)

tree <- groupClade(tree, .node = 44)

## Plot phylogenetic tree:
phylo_tree <- ggtree(tree, aes(color = group,
                               linetype = group)) +
        xlim(0, 2.5) +
        geom_tiplab(size = 4,
                    color = "black",
                    offset = 0.03) +
        geom_label2(aes(subset = !isTip,
                        label = label),
                    size = 3,
                    color = "darkred",
                    alpha = 1) +
        geom_hilight(node = 52,
                     fill = "yellow",
                     alpha = 0.5) +
        theme_tree2("white") +
        scale_color_continuous(low = "grey",
                               high = "grey",
                               name = "Branch length (my)")

## Plot:
plot(phylo_tree) + guides(fill = FALSE)
```
Run lintr:

```{r, message = FALSE}
lintr::lint(file = "./plot_phylogenetic_tree.Rmd")
```

