---
title: "Bombus population genomics"
output: plot_manhattan_for_nsl.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:  
This script takes a tab-delimited file containinng five columns:  
- Column 1: Chromosome (CHR)
- Column 2: Base pair position on chromosome (BP)  
- Column 3: Normalised nsL score per position (P)  
- Column 4: Variant annotation generted by [SnpEff](https://pcingola.github.io/SnpEff/index.html)  
- Column 5: Locus ID  
The script generates a Manhattan plot for normalised raw |nsL| scores.  
The script outputs the plot as a saved R object (.RData), as well as in .png and .pdf format.   

1. Load libraries:  

```{r, message = FALSE}
## Load libraries:
libraries <- c("devtools",
               "ggplot2",
               "PopGenome",
               "qqman",
               "reshape",
               "ggpubr",
               "zoo",
               "stringr",
               "dplyr",
               "ggrepel")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                chbiocLite(pkgs = lib)
                library(lib, character.only = TRUE)
        }
}
```

2. Read in data, format data for Manhattan plot and generate plot:  

```{r, message = FALSE}
## Read in variants:
variants <- read.table(file = "input/nsl_input/input_for_r_annotated.txt",
                       col.names = c("CHR",
                                     "BP",
                                     "P",
                                     "variant",
                                     "locus"))

## Create a column for 'SNP':
variants$SNP <- rownames(variants)
## Ensure nsl scores are absolute values:
variants$P <- abs(variants$P)

## Extract top 50 peaks of interest:
input <- "./input/nsl_input/input_for_plot_top20_ensembl.txt"
peaks <- read.table(file = input,
                    col.names = c("CHR",
                                  "BP",
                                  "P",
                                  "locus"))

peaks_tmp <- subset(variants,
                    BP %in% peaks$BP &
                            P > 2.56)

## Create SNP list of interest:
snps_of_interest <- peaks_tmp$SNP

variants$CHR <- gsub("NC_015762.1", 1, variants$CHR)
variants$CHR <- gsub("NC_015763.1", 2, variants$CHR)
variants$CHR <- gsub("NC_015764.1", 3, variants$CHR)
variants$CHR <- gsub("NC_015765.1", 4, variants$CHR)
variants$CHR <- gsub("NC_015766.1", 5, variants$CHR)
variants$CHR <- gsub("NC_015767.1", 6, variants$CHR)
variants$CHR <- gsub("NC_015768.1", 7, variants$CHR)
variants$CHR <- gsub("NC_015769.1", 8, variants$CHR)
variants$CHR <- gsub("NC_015770.1", 9, variants$CHR)
variants$CHR <- gsub("NC_015771.1", 10, variants$CHR)
variants$CHR <- gsub("NC_015772.1", 11, variants$CHR)
variants$CHR <- gsub("NC_015773.1", 12, variants$CHR)
variants$CHR <- gsub("NC_015774.1", 13, variants$CHR)
variants$CHR <- gsub("NC_015775.1", 14, variants$CHR)
variants$CHR <- gsub("NC_015776.1", 15, variants$CHR)
variants$CHR <- gsub("NC_015777.1", 16, variants$CHR)
variants$CHR <- gsub("NC_015778.1", 17, variants$CHR)
variants$CHR <- gsub("NC_015779.1", 18, variants$CHR)

## Convert CHR to factor and sort levels by correct order:
variants$CHR <- factor(variants$CHR, levels = unique(variants$CHR))
options(scipen = 999)

## Rearrange the order:
variants$P <- abs(variants$P)
variants$SNP <- as.numeric(variants$SNP)

don <- variants %>%
  # Compute chromosome size
  group_by(CHR) %>%
  summarise(chr_len = max(BP)) %>%
  # Calculate cumulative position of each chromosome
  mutate(tot = cumsum(chr_len) - chr_len) %>%
  select(-chr_len) %>%
  # Add this info to the initial dataset
  left_join(variants, ., by = c("CHR" = "CHR")) %>%
  # Add a cumulative position of each SNP
  arrange(CHR, BP) %>%
  mutate(BPcum = BP + tot) %>%
  # Add highlight and annotation information
  mutate(is_highlight = ifelse(SNP %in% snps_of_interest,
                                "yes",
                                "no")) %>%
  mutate(is_annotate = ifelse(P > 2.56,
                             "yes",
                             "no"))
# Prepare X axis
axisdf <- don %>%
        group_by(CHR) %>%
          summarize(center = (max(BPcum) + min(BPcum)) / 2)

# Make the plot
manhattan_plot <- ggplot(don, aes(x = BPcum,
                                  y = P)) +
    # Show all points
    geom_point(aes(color = as.factor(CHR)),
                alpha = 0.3,
                size = .3) +
        ylim(0, 5.5) +
        scale_color_manual(values = rep(c("black",
                                          "navy"),
                                        22)) +
        geom_hline(yintercept = 2.56,
                   color = "blue",
                   linetype = "dashed") +
        geom_hline(yintercept = 2.87,
                   color = "purple",
                   linetype = "dashed") +
        xlab("Chromosome") +
        ylab("|nsl| score") +
        # custom X axis:
        scale_x_continuous(label = axisdf$CHR,
                           breaks = axisdf$center) +
        scale_y_continuous(expand = c(0.01, 0.01)) +
        # Add highlighted points:
        geom_point(data = subset(don,
                                 is_highlight == "yes"),
                   color = "darkorange",
                   alpha = 0.8,
                   size = 1.3) +
        # Add label using ggrepel to avoid overlapping
        geom_label_repel(data = subset(don,
                                       is_highlight == "yes"),
                         aes(label = locus),
                         size = 2.5) +
        # Custom the theme:
        theme_bw() +
        theme(legend.position = "none",
              panel.border = element_blank(),
              panel.grid.major.x = element_blank(),
              panel.grid.minor.x = element_blank(),
              axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title = element_text(size = 15,
                                        face = "bold"))

## Save the plot as an R object:
save(manhattan_plot,
     file = "./results/manhattan_plot.RData")

## Save the plot to file:
ggsave(file = "./results/manhattan_plot.png",
       height = 6,
       width = 10)

## Save as a PDF:
ggsave(file = "./results/manhattan_plot.pdf",
       dpi = 600,
       height = 6,
       width = 10)
```

3. Run lintr:

```{r, message = FALSE}
## Run lintr:
lintr::lint(file = plot_manhattan_for_nsl.Rmd")
```
