---
title: "Bombus population genomics"
output: plot_gggene_arrows.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction
This script takes a tab-delimited file as input containing the following columns:
1) Chromosome name
2) Start position of chromosome
3) End position of chromosome
4) Gene name
5) Strand
6) Status - if you want to highlight a gene of interest
The script generates an arrow plot for a gene of interest indicating the presence of functional domains.  
The script saves the arrow plot as an R object.

1. Load required libraries:  

```{r, message = FALSE}
## Load libraries:
libraries <- c("ggplot2",
               "ggpubr",
               "gggenes",
               "lintr")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE)
        }
}

## Create output directory:
dir.create(path = "results/plots",
           recursive = TRUE)
```

2. Read in input files:

```{r, message = FALSE}
## Read in file for plotting:
input <- "input/gggenes/wolbachia_input_for_gggenes.txt"
data <- read.table(file = input,
                   header = TRUE)
## Assign values to start and end positions:
start_position <- 1649722
end_position   <- 1876527
## Turn off scientific notation:
options(scipen = 999)
data$direction <- ifelse(data$strand == "forward", 1, -1)

data$highlight <- data$start > 1724000 &
                  data$start < 1770171
                  
```

3. Generate arrow plot:

```{r, message = FALSE)
## Plot:
arrow_plot <- ggplot(data,
                     aes(xmin = as.numeric(start),
                         xmax = as.numeric(end),
                         y = strand,
                         fill = highlight,
                         label = gene,
                         forward = direction)) +
        xlim(start_position, end_position) +
        geom_gene_arrow(arrowhead_height = unit(3, "mm"),
                        arrowhead_width = unit(1, "mm")) +
        facet_wrap(~ molecule,
                   scales = "free",
                   nrow = 1,
                   shrink = FALSE) +
        scale_fill_manual(values = c("Grey75",
                                     "Beige")) +
        theme_genes() +
        theme(axis.title.y = element_blank(),
              axis.text.y = element_blank(),
              axis.title.x = element_blank(),
              axis.text.x = element_blank(),
              axis.ticks.x = element_blank(),
              legend.position = "none")

## Save image as an R object:
save(arrow_plot,
       file = "results/plots/plot_gggene_arrows.RData")
```

4. Run lintr to check style:

```{r, message = FALSE}
## Run lintr:
lintr::lint(file = "plot_gggene_arrows.Rmd")
```
