#!/usr/bin/env Rscript
#######################################################################
##
## Author: Joe Colgan           Program: quant_depth.R
##
## Date:   2019/12/02
##
## Purpose:
## This script takes tab-delimited text files generated by samtools
## to extract read depth for user-specific region. It then compares
## the average read depth in this region to the genome-wide average.
## For input, the file takes per sample:
## 1) A tab-delimited text file containing three columns:
## - Chromosome name
## - Position
## - Depth
## 2) A tab-delimited text file containing gene region of interest:
## - Chromosome name
## - Start position for region of interest
## - End position for region of interest
## The user defines the name of output file.
#######################################################################

## Create output directory:
dir.create("results")

## Take arguments from the command line:
## Provide access to a copy of the command line arguments supplied when R session is invoked
args <- commandArgs(TRUE)
region_of_interest <- args[1] 						# Assign first argument as input
depth_file <- args[2]
output <- args[3]						# Assign second argument as input
output_1 <- args[4]
output_2 <- args[5]
message(paste("will output to", output))

## First of all read in gene of interest coordinates:
region_input <- read.table(file = region_of_interest,
header = FALSE)
colnames(region_input) <- c("chromosome",
"gene_start",
"gene_end")
print("Region input loaded")

flank_region_1_start <- region_input$gene_start - 10000
flank_region_1_end <-  region_input$gene_start

flank_region_2_start <- region_input$gene_end                       
flank_region_2_end <-  region_input$gene_end + 10000

## Second read in depth file:
depth_input <- read.table(file = depth_file,
header = FALSE)
## Update column names:
colnames(depth_input) <- c("chromosome",
                           "position",
                           "depth")
## Print to console:
print("Depth input loaded")

## Remove regions of zero counts or putative repetitive regions:
depth_input <- subset(depth_input,
                      depth > 0 &
                      depth < 100)

region_chrom <- as.character(region_input$chromosome)

## Subset depth file using gene coordinates:
depth_subset <- subset(depth_input,
                       chromosome == region_chrom &
                       position >=  region_input$gene_start &
                       position <= region_input$gene_end)
## Print to console:
print("Subset completed")

## Subset depth file for region flanking upstream:
flank_1_subset <- subset(depth_input,
                         chromosome == region_chrom &
                         position >= flank_region_1_start  &
                         position <= flank_region_1_end)

## Subset depth file for region flanking downstream:
flank_2_subset <- subset(depth_input,
                         chromosome == region_chrom &
                         position >= flank_region_2_start &
                         position <= flank_region_2_end)

## Calculate t test between gene of interest and rest of genome:
print(t.test(depth_subset$depth,
             depth_input$depth))
chars <- capture.output(print(t.test(depth_subset$depth,
                                     depth_input$depth)))
## Calculate t test between gene of interest and rest of genome:
print(t.test(flank_1_subset$depth,
             depth_subset$depth))
flank_1_chars <- capture.output(print(t.test(flank_1_subset$depth,
                                             depth_subset$depth)))
## Calculate t test between gene of interest and rest of genome:
print(t.test(flank_2_subset$depth, depth_subset$depth))
flank_2_chars <- capture.output(print(t.test(flank_2_subset$depth,
                                             depth_subset$depth)))

## Output to file:
writeLines(chars,
           con = file(output))
writeLines(flank_1_chars,
           con = file(output_1))
writeLines(flank_2_chars,
           con = file(output_2))

## Print to console:
print("Script complete! Check output")
