---
title: "Bombus population genomics"
output: plot_nsl_by_variant_annotation.html
authors: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:
This script takes input and generates histogram plots as output.  
The first plot generates a histogram plot of variant type by raw |nsl| score.
The second plot generates a histogram plot of max |nsl| score by gene.

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2",
               "ggpubr",
               "lintr")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        avebiocLite(pkgs = lib)
        library(lib, character.only = TRUE)
    }
}
```

First, investigate the relationship between nsl score and variant annotation.

```{r, message = FALSE}
data <- read.table(file = "./input/nsl_input/input_for_r_annotated.txt",
                   col.names = c("chrom",
                                 "position",
                                 "nsl_score",
                                 "annotation",
                                 "locus"))
```

Subdivide SNPs by annotation type:

```{r, message = FALSE}
## Generate plot:
data$annotation_simplified <- data$annotation

## Update levels for plotting:
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "UTR")] <- "UTR"
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "stream")] <- "5kb up/down-stream"
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "intron")] <- "Intron"
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "intergenic")] <- "Intergenic"
levels(data$annotation_simplified)[grep(x = levels(data$annotation_simplified),
                                        pattern = "non_coding_transcript")] <- "Non-coding RNA"
levels(data$annotation_simplified)[!(levels(data$annotation_simplified) %in% c("UTR",
                                                                               "5kb up/down-stream",
                                                                               "Intron",
                                                                               "Intergenic",
                                                                               "Non-coding RNA"))] <- "Coding"

data$annotation_simplified <- factor(data$annotation_simplified,
                                     levels = rev(c("Non-coding RNA",
                                                    "Coding",
                                                    "UTR",
                                                    "Intergenic",
                                                    "5kb up/down-stream",
                                                    "Intron")))
levels(data$annotation_simplified)

## Generate histogram:
variant_histogram <- ggplot(data = data,
                            aes(x = nsl_score,
                                fill = annotation_simplified)) +
        geom_histogram(breaks  = seq(from = 0,
                                     to = 5.3,
                                     by = 0.1)) +
        theme_bw() +
        ylab("Number of SNPs") +
        xlab("|nsl| score") +
        scale_fill_manual(values = c("pink",
                                     "black",
                                     "orange",
                                     "grey",
                                     "brown",
                                     "yellow")) +
        geom_vline(xintercept = 2.56,
                   linetype = "dashed",
                   colour = "blue") +
        annotate("text",
                 label = "Top 1% of genome-wide |nsl| scores",
                 x = 2.46,
                 y = 50000,
                 #angle = 90,
                 colour = "blue",
                 size = 4) +
        annotate("text",
                 label = "Top 10% of genic |nsl| scores",
                 x = 2.97,
                 y = 50000,
                 #angle = 90,
                 colour = "purple",
                 size = 4) +
        geom_vline(xintercept = 2.87,
                   linetype = "dashed",
                   colour = "purple") +
        guides(fill = guide_legend(title = "Variant type")) +
        theme(legend.position = "top") +
        theme(axis.title.x = element_text(size = 15,
                                  face = "bold"),
              axis.text = element_text(size = 10,
                                  face = "plain"),
              axis.title.y = element_blank(),
              legend.text = element_text(size = 10,
                                         face = "plain"),
              legend.title = element_text(size = 10,
                                          face = "bold"),
              legend.justification = c(1.1, 1.1),
              legend.position = c(1, 1)) +
        scale_x_continuous(breaks = c(0:5.5),
                           limits = c(0, 5.5))

## Flip axes
variant_histogram <- variant_histogram +
        coord_flip()

## Save plot:
ggsave(file = "results/histogram_snp_by_variant_type.pdf",
       dpi = 600,
       height = 10,
       width = 12)

## Save object:
save(data = variant_histogram,
     file = "./results/histogram_snp_by_variant_type.RData")
```

Run lintr:

```{r, message = FALSE}
lintr::lint(file = "plot_nsl_by_variant_annotation_modified.Rmd")
```