---
title: "Bombus population genomics"
output: calculate_zscores_gene_density.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:  
The purpose of this script is to examine gene density across the genome with a particular interest in a region of low diversity on chromosome one.  
The script takes a tab-delimited text file containing gene density per 100kb sliding windows, z-score transforms and calculates p values for
genomic regions. The script subsets regions with significantly low and high gene densities.  
The script outputs a tab-delimited text file, as well as an R object.  

1. Load libraries:

```{r, message = FALSE}
## Load libraries:
libraries <- c("lintr",
               "ggplot2",
               "ggpubr")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE)
        }
}

## Create output folder:
dir.create(path = "./results/figures",
           recursive = TRUE)
```

2. Read in input data:

```{r, message = FALSE}
## Read in input file:
data <- read.table(file = "./input/gene_density_100kb_plus_coordinates.txt",
                   header = FALSE)

## Updated column names:
colnames(data) <- c("position",
                    "gene_density")

## Calculate z-score for gene density:
data_gene_density_sd <- sd(data$gene_density) *
                 sqrt((length(data$gene_density) - 1) /
                              (length(data$gene_density)))
data_gene_density_mean <- mean(data$gene_density)

data$gene_density_zscores <- (data$gene_density - data_gene_density_mean) / 
                              data_gene_density_sd
data$gene_density_pvalue2sided <- 2 * pnorm(-abs(data$gene_density_zscores))

summary(data$gene_density_pvalue2sided)
summary(data$gene_density)

## Subset based on significant:
data_sig <- subset(data,
                   gene_density_pvalue2sided < 0.05)
```

3. Write the output to file:

```{r, message = FALSE}
## Turn off scientific notation:
options(scipen = 999)

## write to output
write.table(x = data,
            file = "./results/gene_density_zscore_normalised.txt",
            row.names = FALSE,
            col.names = TRUE,
            quote = FALSE,
            sep = "\t")
```

4. Subset regions with significant high or low nucleotide diversity:

```{r, message = FALSE}
## Save output:
save(data, file = "./results/gene_density_zscore_normalised.Rdata")
```

5. Run lintr:  

```{r, message = FALSE}
lintr::lint(file = "./calculate_zscores_gene_density.Rmd")
```
