---
title: "Bombus population genomics"
output: plot_apis_pca.html
authors: Joe Colgan (joscolgan)
---

```{r}
## Plot PCA of larger number of samples (n=46):
## Assign the input name:  
input_apis_vcf <- "input/apis_mellifera/freebayes_hap0_minQ_1_minaltfrac_0.25_minCov1.Apis_combined.NC_only.minQ20.remove_indels.maxmissing1.0.alts_removed.rare_removed.maxmeanDP_100.recode.vcf"

## Convert VCF to genome data structure (gds) file format
## gds is a container for storing annotation data and SNP genotypes.
## In this format, each byte encodes up to four SNP genotypes thereby reducing file size and access time.
## The GDS format supports data blocking so that only the subset of data that is being processed needs
## to reside in memory. GDS formatted data is also designed for efficient data access to large datasets.

## Use the name of input to designate the name for the GDS file format.
input_apis_vcf_gds <- gsub(".vcf", ".gds", input_apis_vcf)

## Reformat VCF into GDS format
snpgdsVCF2GDS(input_apis_vcf,
              input_apis_vcf_gds,
              method = "biallelic.only")

## Extract genotype information:
input_apis_vcf_gds_genofile <- snpgdsOpen(input_apis_vcf_gds)

## Extract genotype information and generate genotypic matrix:  
input_apis_vcf_gds_genofile_matrix <- read.gdsn(index.gdsn(input_apis_vcf_gds_genofile,
                                                          "genotype"))
                                                          
## For performing PCA, it is recommended to perform LD pruning:
snpset <- snpgdsLDpruning(input_apis_vcf_gds_genofile,
                          ld.threshold = 0.1,
                          autosome.only = FALSE)
                          
## Get all selected SNP ids:
snpset.id <- unlist(unname(snpset))

## Perform PCA and store output in variable:  
pca_apis        <- snpgdsPCA(input_apis_vcf_gds_genofile,
                          num.thread = 20,
                          snp.id = snpset.id,
                          autosome.only = FALSE)

## Calculate percentages for principal components:  
pc_percent_apis      <- pca_apis$varprop*100
## Print rounded up percentages:  
print(round(pc_percent_apis, 2))

## Extract sample names from input:  
sample_id     <- read.gdsn(index.gdsn(input_apis_vcf_gds_genofile,
                                      "sample.id"))

## Read in file containing treatment information. 
## This file should contain 'treatment information' for each sample and be in the same order as each sample. 
## i.e. the first treatment information would be assigned to the first sample, etc., etc.
## Read in table:
pop_code_df <- read.table(file = "input/apis_mellifera/apis_samples.tmp",
                          header = FALSE)

## Rename columns:
colnames(pop_code_df) <- c("sample", "site_number")

##
pop_code_apis <- pop_code_df$site_number

## Generate a dataframe containing a column for:
## 1) Sample name
## 2) Treatment condition
## 3) First principal component of interest
## 4) Second principal component of interest
tab_apis <- data.frame(sample.id = pca_apis$sample.id,
                  pop = factor(pop_code_apis)[match(pca_apis$sample.id,
                                                    sample_id)],
                  EV1 = pca_apis$eigenvect[,1],    # the first eigenvector
                  EV2 = pca_apis$eigenvect[,2],    # the second eigenvector
                  stringsAsFactors = FALSE)

## Unlist as a character string and update sample ids:
tab_apis$sample.id <- pop_code_df$sample

## Add another column for site_id:
tab_apis$site_id <- pop_code_df$site_number

## Alternatively, to plot with just sample ids and not points:
names_plot<- ggplot(tab_apis,
                    aes(x = EV1,
                        y = EV2,
                        color = pop)) +
        theme_minimal() +
        xlab(paste("Principal component", 1," ","(",round(pc_percent_apis[1], 2), "%", ")", sep="")) +
        ylab(paste("Principal component", 2," ","(",round(pc_percent_apis[2], 2), "%", ")", sep="")) + 
        geom_text(size = 6,
                  position = position_jitter(width = 0.04,
                                             height = 0.04),
                  aes(label = site_id,
                      color = pop)) +
        scale_color_manual(values = c("blue",
                                      "orange",
                                      "red",
                                      "grey"))

## Adjust the size of the labels:
names_plot <- names_plot +
                 theme(axis.text = element_text(size = 20),
                 axis.title=element_text(size = 20,
                                         face = "bold")) +
                 theme(legend.position = "none")

ggsave(file = "apis_pca_pruned_0.1_test.png",
       width = 10,
       height = 10)
```
