---
title: "Bombus population genomics"
output: plot_nsl_by_variant_annotation.html
authors: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction:
This script takes input and generates histogram plots as output.  
The first plot generates a histogram plot of variant type by raw |nsl| score.
The second plot generates a histogram plot of max |nsl| score by gene.

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2",
               "ggpubr",
               "lintr",
               "reshape2",
               "gridExtra")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        avebiocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
```

First, investigate the relationship between nsl score and variant annotation.

```{r, message = FALSE}
data <- read.table(file = "./input/variants/input_annotations_per_snp.txt",
                   col.names = c("chrom",
                                 "position",
                                 "nsl_score",
                                 "annotation"))
```

Subdivide SNPs by annotation type:

```{r, message = FALSE}
## Create empty dataframe:
density_df <- setNames(data.frame(matrix(ncol = length(unique(data$annotation)),
                           nrow = 26)),
         c(as.character(unlist(unique(data$annotation)))))

for (variant in unique(data$annotation)){
        print(variant)
        tmp <- subset(data, annotation == variant)
        density_df[variant] <- hist(tmp$nsl_score,
                                    breaks = seq(0, 5.3,
                                                 by = 0.2),
                                    plot = FALSE)$counts
        #density_df[variant] <- density_df[variant]/colSums(density_df[variant])
}

mids <-  hist(tmp$nsl_score,
              breaks = seq(0, 5.3,
                           by = 0.2),
              plot = FALSE)$mids

## Add an additional column with bins to group by in plots:
density_df$block <- factor(mids,
                           levels = c(unique(mids)))

## Melt:
density_melt <- melt(density_df)

sort_order <- names(sort(colSums(density_df[, 1:25]),
           decreasing = TRUE))
colnames(density_melt) = c("bins","variable","value")

levels(density_melt$variable) <- sort_order

##
density_melt$is_intergenic <- density_melt$variable == "intergenic_region"
density_melt$is_coding <- density_melt$variable != "intergenic_region" &
        density_melt$variable != "intronic_variant"

## Turn off scientific notation:
options(scipen = 999)
density_melt$log_value <- log(density_melt$value)

##create a dataframe
ggplot(density_melt,
       aes(x = bins,
               y = log_value,
               fill = is_intergenic)) +
        geom_bar(stat = "identity") +
        labs(fill = "Intergenic") +
        xlab("|nsl| score") +
        ylab("Log2 transformed counts") +
        geom_vline(xintercept = 13.25,
                   linetype = "dashed",
                   colour = "orange") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 15,
                                   hjust = 1),
              axis.title = element_text(size = 15,
                                        face = "bold"),
              legend.position = "bottom",
              legend.title = element_text(size = 15,
                                          face = "bold")) +
        scale_fill_manual(values = c("light blue", "grey"))

## Save plot:
ggsave(file = "./results/histogram_snp_by_variant_type.pdf",
       dpi = 600,
       height = 10,
       width = 12)
```

Plot using only top one percent SNPs:

```{r, message = FALSE}
## Subset values in the top 1%:
updwn_stream_variants <- grep("stream",
                              data$annotation)
updwn_stream_sig_df <- subset(data,
                              rownames(data) %in%
                                      updwn_stream_variants)
updwn_stream_sig_df$variants <- "5kb_up/downstream"

intron_variants <- grep("intron_variant",
                              data$annotation)
intron_variants_df <- subset(data,
                              rownames(data) %in%
                                      intron_variants)
intron_variants_df$variants <- "Intronic"

utr_variants <- grep("UTR",
                              data$annotation)
utr_variants_df <- subset(data,
                     rownames(data) %in% utr_variants)
utr_variants_df$variants <- "UTR"

intergenic_variants <- grep("intergenic",
                              data$annotation)
intergenic_variants_df <- subset(data,
                     rownames(data) %in% intergenic_variants)
intergenic_variants_df$variants <- "Intergenic"

## all non_coding variants:
noncoding <- c(intergenic_variants,
               utr_variants,
               intron_variants,
               updwn_stream_variants)

coding_sites <- subset(data,
          !(rownames(data) %in% noncoding))
coding_sites$variants <- "Coding"

combined_df <- rbind(updwn_stream_sig_df,
                     coding_sites,
                     intron_variants_df,
                     utr_variants_df,
                     intergenic_variants_df)

combined_df$is_intergenic <- NULL
combined_df$is_coding <- NULL
combined_df$chrom <- NULL
combined_df$position <- NULL

## Create empty dataframe:
new_density_df <- setNames(data.frame(matrix(ncol = length(unique(combined_df$variants)),
                           nrow = 53)),
         c(as.character(unlist(unique(combined_df$variants)))))

for (variant in unique(combined_df$variants)){
        print(variant)
        tmp <- subset(combined_df, variants == variant)
        new_density_df[variant] <- hist(tmp$nsl_score,
                                    breaks = seq(0, 5.3,
                                                 by = 0.1),
                                    plot = FALSE)$counts
        #density_df[variant] <- density_df[variant]/colSums(density_df[variant])
}

mids <-  hist(tmp$nsl_score,
              breaks = seq(0, 5.3,
                           by = 0.1),
              plot = FALSE)$mids

## Add an additional column with bins to group by in plots:
new_density_df$block <- factor(mids,
                           levels = c(unique(mids)))

## Melt:
new_density_melt <- melt(new_density_df)

sort_order <- names(sort(colSums(new_density_df[, 1:5]),
           decreasing = TRUE))
colnames(new_density_melt) = c("bins","variable","value")

levels(new_density_melt$variable) <- sort_order

new_density_melt$log_value <- log(new_density_melt$value)

## Remove underscores:
new_density_melt$variable <- gsub("_",
                                  " ",
                                  new_density_melt$variable)

new_density_melt$variable <- factor(new_density_melt$variable,
                                    levels = c(unique(new_density_melt$variable)))

##create a dataframe
scaled_plot <- ggplot(new_density_melt,
       aes(x = bins,
               y = log_value,
               fill = variable)) +
        geom_bar(stat = "identity") +
        labs(fill = "Variant type") +
        xlab("|nsl| score") +
        ylab("Log2 transformed counts") +
        geom_vline(xintercept = 26.25,
                   linetype = "dashed",
                   colour = "black") +
        theme_bw() +
        theme(axis.text = element_text(size = 10,
                                       face = "plain"),
              axis.title = element_text(size = 15,
                                        face = "bold"),
              legend.position = "left",
              legend.title = element_text(size = 15,
                                          face = "bold")) +
        scale_fill_manual(values = c("orange",
                                     "yellow",
                                     "brown",
                                     "grey",
                                     "pink")) +
        scale_x_discrete(breaks = levels(new_density_melt$bins)[c(T,
                                                                  rep(F, 5))])

##create a dataframe
raw_plot <- ggplot(new_density_melt,
       aes(x = bins,
               y = value,
               fill = variable)) +
        geom_bar(stat = "identity") +
        labs(fill = "Variant type") +
        xlab("|nsl| score") +
        ylab("Number of polymorphic sites") +
        geom_vline(xintercept = 26,
                   linetype = "dashed",
                   colour = "orange") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 15,
                                   hjust = 1),
              axis.title = element_text(size = 15,
                                        face = "bold"),
              #legend.position = "left",
              legend.title = element_text(size = 15,
                                          face = "bold"),
              legend.text = element_text(size = 10,
                                         face = "plain"),
              legend.justification = c(1.1, 1.1),
              legend.position = c(1, 1)) +
        scale_fill_manual(values = c("orange",
                                     "yellow",
                                     "brown",
                                     "grey",
                                     "pink")) +
                scale_x_discrete(breaks = levels(new_density_melt$bins)[c(T,
                                                                            rep(F, 5))])

ggsave(file = "./results/figures/variant_type.png",
       height = 10,
       width = 7)

## Save as an R object:
save(scaled_plot,
     file = "results/figures/nsl_variant.Rdata")

```

Plot by gene:

```{r, message = FALSE}
## Read in nsl score per gene for plotting:
genes <- read.table(file = "./input/variants/input_for_R.tmp",
                    col.names = c("locus",
                                  "gene_length",
                                  "median_nsl_score",
                                  "nsl_score",
                                  "go_terms"))

## Sort by max absolute nsl score per gene:
genes <- genes[order(-genes$nsl_score), ]

## Subset top 1%:
top_one_percent <- subset(genes,
                          genes$nsl_score > quantile(data$nsl_score,
                                                     0.99))
## Plot:
gene_histogram <- ggplot(data = genes,
                        aes(x = nsl_score),
                        color = "black") +
        geom_histogram(bins = 100,
                       aes(x = nsl_score),
                       fill = "dark blue",
                       colour = "black") +
        geom_histogram(data = top_one_percent,
                       bins = 100,
                       aes(x = nsl_score),
                       fill = "light blue") +
        theme_bw() +
        ylab("Number of genes") +
        xlab("|nsl| score") +
        scale_color_manual(values = c("dark blue",
                                      "light blue")) +
        theme(axis.title = element_text(size = 20,
                                  face = "bold"),
              axis.text = element_text(size = 15,
                                  face = "plain"))

## Save plot to file:
ggsave(file = "./results/histogram_gene_by_nsl_score.pdf",
       dpi = 600,
       height = 8,
       width = 8)

## Save object:
save(data = gene_histogram,
     file = "./results/histogram_gene_by_nsl_score.Rdata")
```

Combine the SNP and gene histograms:

```{r, message = FALSE}
## Generate a combined plot:
ggarrange(variant_histogram,
          gene_histogram,
          labels = c("A", "B"),
          ncol = 1,
          nrow = 2,
          align = "hv")

## Save to file:
ggsave(file = "./results/combined_test_plot.png",
       height = 10,
       width = 10)
```

Run lintr:

```{r, message = FALSE}
lintr::lint(file = "./plot_nsl_by_variant_annotation.Rmd")
```